<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Creating RDML file from raw data • RDML</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">RDML</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kablag/RDML">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Creating RDML file from raw data</h1>
                        <h4 class="author">Konstantin A. Blagodatskikh, Michał Burdukiewicz, Stefan Rödiger</h4>
            
            <h4 class="date">2017-07-23</h4>
          </div>

    
    
<div class="contents">
<div id="creating-rdml-file-from-raw-data" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-rdml-file-from-raw-data" class="anchor"></a>Creating RDML file from raw data</h1>
<p>This vignette describes important step how to create <strong>RDML</strong> object from user provided data and calculate Cq values with custom <strong>RDML</strong> class method. We use a real-world example to demonstrate the use of the <strong>RDML</strong> package.</p>
<p>For this purpose we used data from the <em>VideoScan</em> platform. The <em>VideoScan</em> platform is a highly versatile fluorescence microscope based imaging platform for the real-time analysis of biomolecular interactions. <em>VideoScan</em> has been extended by a heating/cooling-unit, which can be used for DNA melting curve analysis <span class="citation">[@roediger_RJ_2013]</span>, real-time quantitative PCR <span class="citation">[@roediger_highly_2013]</span> and quantitative isothermal amplification <span class="citation">[@spiess_impact_2014]</span>. In this example we focus on qPCR data obtained from the amplification of the human gene <em>mlc-2v</em> <span class="citation">[@sheikh_2015]</span>. The experimental details (e.g., primer sequences, amplification conditions) are described in <span class="citation">[@roediger_highly_2013; @george_development_2016]</span>.</p>
<p>In detail, this dataset contains three amplification curves collected by three runs monitored with TaqMan qPCR chemistry. Similar to other experimental platforms has the <em>VideoScan</em> systems an output as <em>csv</em> file (comma separated value). In general, <em>csv</em> is a format that is sufficient for an in house usage. However, the exchange with other scientists is demanding, since experimental metadata is not consistently included in csv files. These are for example information about qPCR conditions, primers, experimentalist and so on as described in the MIQE and RDML standard documentations.</p>
<p>All <em>csv</em> raw data were transferred to the <strong>R</strong> environment and stored as <em>rda</em> files in the <strong>C54</strong> dataset after completion of the experiment as described elsewhere <span class="citation">[@RDCT2010c]</span>. The <strong>R</strong> environment provides numerous packages for the import of standard file formats, including formats from spread sheet applications (e.g., <em>XLS</em>, <em>XLSX</em>, <em>ODS</em>) <span class="citation">[@Schutten_2014_ods, @Warnes_2015_gdata, @dragulescu_xlsx_2014]</span>, an ODBC connection <span class="citation">[@ripley_rodbc_2015]</span> or from the clipboard (<code>read.table("clipboard")</code>). There are several graphical user interfaces (GUIs) for <strong>R</strong>, which facilitate an easy import of raw data <span class="citation">[@Valero_2012]</span>. For example <strong>RKWard</strong> <span class="citation">[@roediger_rkward_2012]</span> has import functionality for <em>csv</em>, <em>txt</em>, <em>spss</em> and <em>stata</em> files.</p>
<p>We will use the <strong>C54</strong> dataset directly from the <a href="https://CRAN.R-project.org/package=chipPCR">chipPCR</a> package <span class="citation">[@roediger2015chippcr]</span>. The amplification curves have different number of cycles and one of them contains a missing value due to a sensor dropout. Thus, this data should be preprocessed before Cq calculation.</p>
<p><strong>R</strong> is a language with dynamic typing. This is helpful while scripting, but can lead to problematic debugging of more complicated workflows. <strong><em>R6</em></strong> classes provide type-safe interfaces to set data without access to the inner structure of objects so that all imported data can be validated. This option is highly useful when creating packages at an intermediate level for other packages (e.g., such an approach does not permit set type <strong><em>character</em></strong> in place of type <strong><em>integer</em></strong>). Furthermore, the inheritance of <strong><em>R6</em></strong> objects unifies the structure of the package and streamlines the extending of its capabilities (since the whole package is written around a single base class).</p>
<div id="initialization" class="section level2">
<h2 class="hasAnchor">
<a href="#initialization" class="anchor"></a>Initialization</h2>
<p>First of all we have to load all necessary packages and fluorescence data into our <strong>R</strong> session.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RDML)
<span class="kw">library</span>(chipPCR)
<span class="kw">library</span>(MBmca)
<span class="kw">library</span>(ggplot2)
<span class="kw">data</span>(C54)</code></pre></div>
<p>We can see that our data are arranged in data.frame with cycle numbers in the first column and fluorescence values in the other columns. Such data structure is ready to use with the <strong>RDML</strong> package and does not need any conversion.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(C54)
<span class="co">#&gt; 'data.frame':    56 obs. of  4 variables:</span>
<span class="co">#&gt;  $ Cycle: num  0 1 2 3 4 5 6 7 8 9 ...</span>
<span class="co">#&gt;  $ D1   : num  0.141 0.146 0.152 0.151 0.153 ...</span>
<span class="co">#&gt;  $ D2   : num  0.148 0.148 0.148 0.149 0.149 ...</span>
<span class="co">#&gt;  $ D3   : num  0.145 0.147 0.148 0.148 0.149 ...</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
dat &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Cycle =</span> <span class="kw">rep</span>(C54[, <span class="dv">1</span>], <span class="kw">ncol</span>(C54) -<span class="st"> </span><span class="dv">1</span>),
                  <span class="dt">Experiment =</span> <span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">colnames</span>(C54)[-<span class="dv">1</span>], function(i) <span class="kw">rep</span>(i, <span class="kw">nrow</span>(C54)))),
                  <span class="dt">refMFi =</span> <span class="kw">unlist</span>(C54[, -<span class="dv">1</span>]))
<span class="kw">levels</span>(dat[[<span class="st">"Experiment"</span>]]) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"D1 - Stock cDNA"</span>, <span class="st">"D2 - 1/10 cDNA"</span>, <span class="st">"D3 - 1/100 cDNA"</span>)

<span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="dt">x =</span> Cycle, <span class="dt">y =</span> refMFi, <span class="dt">colour =</span> Experiment)) +<span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="CreateRDML_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
</div>
<div id="creating-custom-method-for-rdml-class" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-custom-method-for-rdml-class" class="anchor"></a>Creating custom method for <strong>RDML</strong> class</h2>
<p>The preprocessing and Cq calculation steps are held within a <strong>RDML</strong> object. To do this we shall add custom method to the <strong>RDML</strong> class and use this advanced class as <strong>RDML</strong> object initializator (data import).</p>
<p>As described previously <span class="citation">[@roediger2015r]</span>, we first preprocess the raw data with the <strong>CPP</strong> function from the <a href="https://CRAN.R-project.org/package=chipPCR">chipPCR</a> package. Next, these preprocessed data are stored in new runs within the same experiment. For Cq calculation we used the <strong>diffQ2</strong> function from the <a href="https://CRAN.R-project.org/package=MBmca">MBmca</a> <span class="citation">[@roediger_RJ_2013]</span> package. Note, there are other functions available within <strong>R</strong> to determine the Cq values. The <a href="https://CRAN.R-project.org/package=qpcR">qpcR</a> package is one of the most popular packages for such tasks.</p>
<p>Additional our method will have two arguments:</p>
<ul>
<li>vector <code>last.cycle</code> – cycle limits for every amplification curve;</li>
<li>list <code>bg.range</code> – background range for every amplification curve.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">RDML$<span class="kw">set</span>(<span class="st">"public"</span>, <span class="st">"ProcessVideoScan"</span>,
         function(last.cycle,
                  bg.range) {
           <span class="co"># Get curves description</span>
           tab &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(self$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>())
           <span class="co"># Get fluorescent point</span>
           dat &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(self$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(tab))
           <span class="co"># Give new names to runs</span>
           <span class="co"># Preprocess amplification curve raw data as described in </span>
           <span class="co"># R&amp;ouml;diger et al. (2015) Bioinformatics. Note that the dataset</span>
           <span class="co"># has different length of cycle numbers and missing values (NAs).</span>
           <span class="co"># The CPP function from the chipPCR package prepares the data for </span>
           <span class="co"># further analysis.</span>
           tab[, <span class="st">"run.id"</span>] &lt;-<span class="st"> </span><span class="kw">paste0</span>(tab[, <span class="st">"run.id"</span>], <span class="st">"_CPP"</span>)
           
           <span class="co"># Curves will be used one by one</span>
           for(i in <span class="dv">2</span>:<span class="kw">length</span>(dat)) {
             <span class="co"># Preprocess data</span>
             preprocessed &lt;-<span class="st"> </span><span class="kw">CPP</span>(dat[<span class="dv">1</span>:last.cycle[i -<span class="st"> </span><span class="dv">1</span>], <span class="dv">1</span>], 
                                 dat[<span class="dv">1</span>:last.cycle[i -<span class="st"> </span><span class="dv">1</span>], i],
                                 <span class="dt">trans =</span> <span class="ot">TRUE</span>,
                                 <span class="dt">bg.range =</span> bg.range[[i -<span class="st"> </span><span class="dv">1</span>]])[[<span class="st">"y.norm"</span>]]
             <span class="co"># Create data.frame with cycle number and preprocessed curve Then</span>
             <span class="co"># give name to fluorescence points columns as before preprocessing</span>
             dat_CPP &lt;-<span class="st"> </span><span class="kw">cbind</span>(dat[<span class="dv">1</span>:last.cycle[i -<span class="st"> </span><span class="dv">1</span>], <span class="dv">1</span>],
                              preprocessed)
             <span class="kw">colnames</span>(dat_CPP)[<span class="dv">2</span>] &lt;-<span class="st"> </span>tab$fdata.name[i -<span class="st"> </span><span class="dv">1</span>]
             <span class="co"># Set preprocessed data with new description (new run names)</span>
             self$<span class="kw"><a href="../reference/setfdata-function.html">SetFData</a></span>(dat_CPP, tab)
             <span class="co"># Calculate and set Cq</span>
             <span class="co"># Set Cq from second derivative maximum method as described in </span>
             <span class="co"># R&amp;ouml;diger et al. (2015) Bioinformatics for preprocessed data.</span>
             <span class="co"># The diffQ2 function from the MBmca package </span>
             <span class="co"># (R&amp;ouml;diger et al. (2013), The R Journal) was used to calculate the</span>
             <span class="co"># Cq values of each amplification curve.</span>
             cq &lt;-<span class="st"> </span><span class="kw">diffQ2</span>(dat_CPP, <span class="dt">inder =</span> <span class="ot">TRUE</span>)[[<span class="st">"xTm1.2.D2"</span>]][<span class="dv">1</span>]
             self$experiment[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"exp.id"</span>]]]$
<span class="st">               </span>run[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"run.id"</span>]]]$
<span class="st">               </span>react[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"react.id"</span>]]]$
<span class="st">               </span>data[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"target"</span>]]]$cq &lt;-<span class="st"> </span>cq
           }
         }, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>
)</code></pre></div>
</div>
<div id="rdml-object-creation" class="section level2">
<h2 class="hasAnchor">
<a href="#rdml-object-creation" class="anchor"></a><strong>RDML</strong> object creation</h2>
<p>Now we can create an empty <strong>RDML</strong> object with our advanced class and set the fluorescence data into it. For setting the data we have to create a description (metadata) of the amplification curves, which contain:</p>
<ul>
<li>
<code>fdata.name</code> – curve name. Have to be as columns names in data.frame with fluorescence data.</li>
<li>
<code>exp.id</code> – experiment name.</li>
<li>
<code>run.id</code> – run name. In our case we have three runs.</li>
<li>
<code>react.id</code> – react ID. E.g., number of tube within a reaction plate.</li>
<li>
<code>sample</code> – sample name.</li>
<li>
<code>type</code> – sample type.</li>
<li>
<code>target</code> – target name. E.g., gene of interest name.</li>
<li>
<code>target.dyeID</code> – fluorescent dye used for detection.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create a data frame of metadata </span>
description &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">fdata.name =</span> <span class="kw">c</span>(<span class="st">"D1"</span>, <span class="st">"D2"</span>, <span class="st">"D3"</span>),
  <span class="dt">exp.id =</span> <span class="kw">c</span>(<span class="st">"exp1"</span>, <span class="st">"exp1"</span>, <span class="st">"exp1"</span>),
  <span class="dt">run.id =</span> <span class="kw">c</span>(<span class="st">"run1"</span>, <span class="st">"run2"</span>, <span class="st">"run3"</span>),
  <span class="dt">react.id =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>),
  <span class="dt">sample =</span> <span class="kw">c</span>(<span class="st">"D1 - Stock cDNA"</span>, <span class="st">"D2 - 1/10 cDNA"</span>, <span class="st">"D3 - 1/100 cDNA"</span>),
  <span class="dt">sample.type =</span> <span class="kw">c</span>(<span class="st">"unkn"</span>, <span class="st">"unkn"</span>, <span class="st">"unkn"</span>),
  <span class="dt">target =</span> <span class="kw">c</span>(<span class="st">"MLC-2v"</span>, <span class="st">"MLC-2v"</span>, <span class="st">"MLC-2v"</span>),
  <span class="dt">target.dyeId =</span> <span class="kw">c</span>(<span class="st">"Cy5"</span>, <span class="st">"Cy5"</span>, <span class="st">"Cy5"</span>),
  <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>
)

<span class="co"># Create an empty RDML object</span>
video.scan &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>()

<span class="co"># Add fluorescence data and metadata to the RDML object from a given source.</span>
<span class="co"># For the sake of simplicity we use the C54 dataset from the chipPCR package.</span>
video.scan$<span class="kw"><a href="../reference/setfdata-function.html">SetFData</a></span>(C54, description)</code></pre></div>
</div>
<div id="setting-additional-information" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-additional-information" class="anchor"></a>Setting additional information</h2>
<p>As optional step, we can add some information about our experiment. Such as experimenter contacts, references to documentation, description of samples and targets, thermal cycling conditions, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Add experimentalist information</span>
video.scan$experimenter &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">list</span>(
    experimenterType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"SR"</span>),
      <span class="st">"Stefan"</span>,
      <span class="st">"Roediger"</span>,
      <span class="st">"stefan.roediger@b-tu.de"</span>
    ),
    experimenterType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"CD"</span>),
      <span class="st">"Claudia"</span>,
      <span class="st">"Deutschmann"</span>
    )
  )

<span class="co"># Add a reference to documentation</span>
video.scan$documentation &lt;-<span class="st"> </span><span class="kw">list</span>(
  documentationType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Roediger et al. 2013"</span>),
    <span class="kw">paste</span>(<span class="st">"A Highly Versatile Microscope Imaging Technology Platform for the Multiplex"</span>,
          <span class="st">"Real-Time Detection of Biomolecules and Autoimmune Antibodies. S. Roediger,"</span>,
          <span class="st">"P. Schierack, A. Boehm, J. Nitschke, I. Berger, U. Froemmel, C. Schmidt, M."</span>,
          <span class="st">"Ruhland, I. Schimke, D. Roggenbuck, W. Lehmann and C. Schroeder. Advances in"</span>,
          <span class="st">"Biochemical Bioengineering/Biotechnology. 133:33-74, 2013."</span>,
          <span class="st">"https://www.ncbi.nlm.nih.gov/pubmed/22437246"</span>)
  )
)

cdna &lt;-<span class="st"> </span>
<span class="st">  </span>cdnaSynthesisMethodType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    <span class="dt">enzyme =</span> <span class="st">"SuperScript II"</span>,
    <span class="dt">primingMethod =</span>
      primingMethodType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"oligo-dt"</span>),
    <span class="dt">dnaseTreatment =</span> <span class="ot">TRUE</span>
  )
video.scan$sample$<span class="st">`</span><span class="dt">D1 - Stock cDNA</span><span class="st">`</span>$description &lt;-<span class="st"> "Input stock cDNA was used undiluted (D1)"</span>
video.scan$sample$<span class="st">`</span><span class="dt">D1 - Stock cDNA</span><span class="st">`</span>$cdnaSynthesisMethod &lt;-<span class="st"> </span>cdna
video.scan$sample$<span class="st">`</span><span class="dt">D2 - 1/10 cDNA</span><span class="st">`</span>$description &lt;-<span class="st"> "1/1000 diluted in A. bidest"</span>
video.scan$sample$<span class="st">`</span><span class="dt">D2 - 1/10 cDNA</span><span class="st">`</span>$cdnaSynthesisMethod &lt;-<span class="st"> </span>cdna
video.scan$sample$<span class="st">`</span><span class="dt">D3 - 1/100 cDNA</span><span class="st">`</span>$description &lt;-<span class="st"> "1/1000000 diluted in A. bidest"</span>
video.scan$sample$<span class="st">`</span><span class="dt">D3 - 1/100 cDNA</span><span class="st">`</span>$cdnaSynthesisMethod &lt;-<span class="st"> </span>cdna

video.scan$target$<span class="st">`</span><span class="dt">MLC-2v</span><span class="st">`</span>$xRef &lt;-<span class="st"> </span><span class="kw">list</span>(
  xRefType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"uniprot"</span>,
               <span class="st">"P10916"</span>)
)
video.scan$target$<span class="st">`</span><span class="dt">MLC-2v</span><span class="st">`</span>$sequences &lt;-<span class="st"> </span>
<span class="st">  </span>sequencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    forwardPrimer &lt;-<span class="st"> </span>oligoType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      <span class="dt">sequence =</span> <span class="st">"ACAGGGATGGCTTCATTGAC"</span>),
    reversePrimer &lt;-<span class="st"> </span>oligoType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      <span class="dt">sequence =</span> <span class="st">"ATGCGTTGAGAATGGTTTCC"</span>),
    probe1 &lt;-<span class="st"> </span>oligoType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      <span class="dt">threePrimeTag =</span> <span class="st">"Atto647N"</span>,
      <span class="dt">sequence =</span> <span class="st">"CAGGGTCCGCTCCCTTAAGTTTCTCC"</span>,
      <span class="dt">fivePrimeTag =</span> <span class="st">"BHQ2"</span>)
  )

tcc &lt;-<span class="st"> </span>
<span class="st">  </span>thermalCyclingConditionsType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>),
    <span class="dt">experimenter =</span> <span class="kw">list</span>(
      idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"SR"</span>),
      idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"CD"</span>)
    ),
    <span class="dt">step =</span> 
      <span class="kw">list</span>(
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">1</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dv">95</span>,
                                            <span class="dv">600</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">2</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dv">95</span>,
                                            <span class="dv">40</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">3</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="fl">58.5</span>,
                                            <span class="dv">90</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">4</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="fl">68.5</span>,
                                            <span class="dv">90</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">5</span>,
          <span class="dt">loop =</span> loopType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dt">goto =</span> <span class="dv">2</span>,
                              <span class="dt">repeat.n =</span> <span class="dv">49</span>)
        )
      )
  )
video.scan$thermalCyclingConditions &lt;-<span class="st"> </span><span class="kw">list</span>(
  tcc
)

<span class="co">#add description of the experiment</span>
video.scan$experiment$exp1$description &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">paste</span>(<span class="st">"The aim was to amplify MLC-2v in the VideoScan and to monitor with a"</span>,
        <span class="st">"hydrolysis probe for MLC-2v. The primer sequences for MLC-2v were taken"</span>,
        <span class="st">"from Roediger et al. (2013). The amplification was detected in solution of"</span>,
        <span class="st">"the 1 HCU (see Roediger et al. 2013 for details). A 20 micro L PCR reaction"</span>,
        <span class="st">"was composed of 250 nM primer (forward and reverse), 1x Maxima Probe qPCR"</span>,
        <span class="st">"Master Mix (Fermentas), 1 micro L template (MLC-2v amplification product in"</span>,
        <span class="st">"different dilutions), 50 nM hydrolysis probe for MLC-2v and A."</span>,
        <span class="st">"bidest. During the amplification, fluorescence was measured at 59.5 degree"</span>,
        <span class="st">"Celsius. The Cy5 channel was used to monitor the MLC-2v specific hydrolysis"</span>,
        <span class="st">"probe. Input stock cDNA was used undiluted (D1). D2 was 1/1000 and D3"</span>,
        <span class="st">"1/1000000 diluted in A. bidest. The D1, D2, and D3 have different numbers"</span>,
        <span class="st">"measure points and D2 contains a missing value at cycle 37."</span>)
video.scan$experiment$exp1$run$run1$thermalCyclingConditions &lt;-<span class="st"> </span>idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>)
video.scan$experiment$exp1$run$run2$thermalCyclingConditions &lt;-<span class="st"> </span>idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>)
video.scan$experiment$exp1$run$run3$thermalCyclingConditions &lt;-<span class="st"> </span>idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>)</code></pre></div>
</div>
<div id="processing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#processing-data" class="anchor"></a>Processing data</h2>
<p>After creating <strong>RDML</strong> object and setting data we can use our method <code>$ProcessVideoScan</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Process VideoScan data</span>
video.scan$<span class="kw">ProcessVideoScan</span>(<span class="dt">last.cycle =</span> <span class="kw">c</span>(<span class="dv">35</span>, <span class="dv">45</span>, <span class="dv">55</span>),
                            <span class="dt">bg.range =</span> <span class="kw">list</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8</span>),
                                           <span class="ot">NULL</span>,
                                           <span class="ot">NULL</span>))
<span class="co">#&gt; Warning in lm.coefs(x[BG], y[BG], method.reg): Chosen method lmrob failed to converge. </span>
<span class="co">#&gt;                    Performed linear regression.</span></code></pre></div>
<p>Then visualize the object with the <code>$AsDendrogram</code> method (runs with IDs that include <code>_CPP</code> in names, contain the preprocessed curves):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualize RDML object</span>
video.scan$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="CreateRDML_files/figure-html/unnamed-chunk-9-1.png" width="576"></p>
<p>And plot preprocessed curves indicating the Cq values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Visualise preprocessed data with Cq values as vertical dashed lines
<span class="co"># Add custom column that contains the calculated Cq</span>
tab &lt;-<span class="st"> </span>video.scan$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>(<span class="dt">cq =</span> {
                 cq &lt;-<span class="st"> </span>data$cq
                 if (<span class="kw">is.null</span>(cq) ||<span class="st"> </span><span class="kw">is.na</span>(cq))
                   <span class="ot">NULL</span>
                 else
                   cq
               })
<span class="co"># Get preprocessed data in 'long.table' format</span>
dat &lt;-<span class="st"> </span>video.scan$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(tab[<span class="kw">grepl</span>(<span class="st">"_CPP"</span>, tab[[<span class="st">"run.id"</span>]]), ],
                           <span class="dt">long.table =</span> <span class="ot">TRUE</span>)
<span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="dt">x =</span> cyc, <span class="dt">y =</span> fluor)) +
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">group =</span> fdata.name, <span class="dt">color =</span> fdata.name),
            <span class="dt">size =</span> <span class="fl">1.2</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="kw">aes</span>(<span class="dt">xintercept =</span> cq, <span class="dt">color =</span> fdata.name),
             <span class="dt">linetype =</span> <span class="st">"longdash"</span>) +
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="st">"Cycle"</span>) +
<span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="st">"Fluorescence"</span>) +
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Amplification of human MLC-2v</span><span class="ch">\n</span><span class="st">VideoScan HCU"</span>)</code></pre></div>
<p><img src="CreateRDML_files/figure-html/unnamed-chunk-10-1.png" width="576"></p>
<p>However, there are further <strong>R</strong> packages on <a href="https://cran.r-project.org/">CRAN</a>, <a href="http://bioconductor.org/">Bioconductor</a> and <a href="https://github.com/">GitHub</a>, which can be used with the RDML package <span class="citation">[@roediger2015r]</span> to deal with non-detects <span class="citation">[@mccall_non-detects_2014]</span>, normalization <span class="citation">[@perkins_readqpcr_2012]</span>, expression analysis <span class="citation">[@dvinge_htqpcr:_2009, matz_no_2013]</span> and periodicity in qPCR systems <span class="citation">[@spiess_system-specific_2016]</span>.</p>
</div>
<div id="export-to-file" class="section level2">
<h2 class="hasAnchor">
<a href="#export-to-file" class="anchor"></a>Export to file</h2>
<p>To save our <strong>RDML</strong> object as file we can use <code>$AsXML()</code> method with file name as argument:</p>
<p><code>video.scan$AsXML("filename.RDML")</code></p>
<p><em>Note that currently due to limits of <code>XML</code> package it is rather slow operation!!!</em></p>
<p>A XML files can be compressed to save space. The <strong>RMDL</strong> package uses the <code>zip()</code> function from the <strong>utils</strong> package with default values to compress the RDML file. Therefore, the compression mode depends on the operating system. The <strong>RDML</strong> function <code>AsXML()</code> without an argument returns one string, which contains the XML data. This can be used to compress it with any other compression algorithm.</p>
</div>
<div id="benchmark-of-the-rdml-generation" class="section level2">
<h2 class="hasAnchor">
<a href="#benchmark-of-the-rdml-generation" class="anchor"></a>Benchmark of the RDML generation</h2>
<p>The XML format becomes less efficient to work with at larger file sizes and complex structured information (e.g., matrices). Therefore, a limitation of RMDL is the process of the file generation. This characteristic is inherited by the <strong>RDML</strong> package. Speed is an issue in high-throughput applications. To give the user a foundation for a decision support if the implementation of the RMDL format is advisable, we performed a benchmark. We used the <strong>microbenchmark</strong> package <span class="citation">[@Mersmann_2014_microbenchmark, @Radford_2014_microbenchmark]</span> to assess the time required for generation of RDML files. The simulation included datasets with ranging from 1 to 1000 entries per file.</p>
</div>
<div id="dealing-with-other-data-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#dealing-with-other-data-formats" class="anchor"></a>Dealing with other data formats</h2>
<p>Researchers often collect gene expression data from more than one laboratory and and are thus required to analyze and aggregate many data sets. Adaptable data management - also known as “adaptive informatics” - is relevant in cases where data from different omics approaches and assays (e.g., flow cytometry, digital PCR, NGS) need to be merged <span class="citation">[@baker_quantitative_2012]</span>. As RDML is based on XML, it is possible to converge the files with other formats such as HDF, <span class="citation">[@millard_adaptive_2011]</span> which is supported by <strong>R</strong> <span class="citation">[@Fischer_HDF5]</span>. This enables extended data storage and analysis, but also a higher level of experimental data management.</p>
</div>
<div id="conclusion" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h2>
<p>In this example we used raw data from the <em>VideoScan</em> platform to demonstrate the creation of an <em>RDML</em> file via the <strong>RDML</strong> package. This procedure can be easily adapted to other experiments and devices and gives the researcher a tool to create a file in a standardized format for convenient data exchange. In contrast to simple <em>csv</em> files contains this file additional metadata in a defined structure. This in turn makes the data available to other scientists for reproducible research.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#creating-rdml-file-from-raw-data">Creating RDML file from raw data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#initialization">Initialization</a></li>
      <li><a href="#creating-custom-method-for-rdml-class">Creating custom method for <strong>RDML</strong> class</a></li>
      <li><a href="#rdml-object-creation"><strong>RDML</strong> object creation</a></li>
      <li><a href="#setting-additional-information">Setting additional information</a></li>
      <li><a href="#processing-data">Processing data</a></li>
      <li><a href="#export-to-file">Export to file</a></li>
      <li><a href="#benchmark-of-the-rdml-generation">Benchmark of the RDML generation</a></li>
      <li><a href="#dealing-with-other-data-formats">Dealing with other data formats</a></li>
      <li><a href="#conclusion">Conclusion</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Konstantin A. Blagodatskikh, Stefan Roediger, Michal Burdukiewicz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
