<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the RDML package • RDML</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">RDML</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kablag/RDML">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Using the RDML package</h1>
                        <h4 class="author">Konstantin A. Blagodatskikh, Michał Burdukiewicz, Stefan Rödiger</h4>
            
            <h4 class="date">2017-07-25</h4>
          </div>

    
    
<div class="contents">
<div id="introduction-to-the-rdml-package" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-the-rdml-package" class="anchor"></a>Introduction to the RDML package</h2>
<p>The <a href="https://CRAN.R-project.org/package=RDML">RDML</a> package was created to work with the Real-time PCR Data Markup Language (RDML) – a structured and universal data standard for exchanging quantitative PCR (qPCR) data <span class="citation">(Lefever et al. 2009; Jan M Ruijter et al. 2015)</span>. RDML belongs to the family of eXtensible Markup Languages (XML), and contains fluorescence data as well as information about the qPCR experiment. A description and the RDML schema and the RDML format are available at <a href="http://rdml.org" class="uri">http://rdml.org</a>.</p>
<p>The XML technology is commonly used in the <strong>R</strong> environment and there are several tools to manipulate XML files <span class="citation">(Nolan and Lang 2014)</span>. When working with XML data, the workhorse package is <strong>XML</strong> <span class="citation">(Lang and CRAN Team 2015)</span>. Therefore, investments in development, implementation and maintenance is moderate.</p>
<p>We use the <strong>R6</strong> <span class="citation">(Chang 2015)</span>, <strong>assertthat</strong> <span class="citation">(Wickham 2013)</span>, <strong>plyr</strong> <span class="citation">(Wickham 2011)</span>, <strong>dplyr</strong> <span class="citation">(Wickham and Francois 2015)</span>, <strong>tidyr</strong> <span class="citation">(Wickham 2014)</span> and <strong>rlist</strong> <span class="citation">(Ren 2015)</span> packages.</p>
<div id="philosphy-of-the-rdml-package" class="section level3">
<h3 class="hasAnchor">
<a href="#philosphy-of-the-rdml-package" class="anchor"></a>Philosphy of the RDML package</h3>
<p><code>RDML</code> imports various data formats (CSV, XMLX) apart from the RDML format. Provided that the raw data have a defined structure (as described in the vignette), the import should be done by a few clicks. The example below shows the import of amplification curve data, which were stored in a CSV file. The function <code><a href="../reference/rdmlEdit.html">rdmlEdit()</a></code> was used in the <a href="https://rkward.kde.org/">RKWard IDE/GUI</a> <span class="citation">(Rödiger et al. 2012)</span> (tested with version 0.6.9z+0.7.0+devel1 on Kubuntu 17.04; NOTE: problems were reported on systems where not the webkit component was used for the rendering of the <code><a href="../reference/rdmlEdit.html">rdmlEdit()</a></code> GUI) for further processing.</p>
<p><img src="File_import.png" alt="HTML5 Icon" style="width:663px;height:332px"></p>
<p>Once imported, this enables <code>rdmlEDIT()</code> and other functions from the <code>RDML</code> package to conduct complex data visualization and processing in the R statistical computing environment.</p>
<p><img src="data_view.png" alt="HTML5 Icon" style="width:663px;height:332px"></p>
</div>
<div id="major-functionalities-of-the-rdml-package" class="section level3">
<h3 class="hasAnchor">
<a href="#major-functionalities-of-the-rdml-package" class="anchor"></a>Major functionalities of the RDML package</h3>
<p>The public methods of the main R6 “RDML” class can be used to access and process the internally stored RDML data. These methods include:</p>
<ul>
<li>
<strong>new()</strong> – creates a new RDML object, empty or from a specified RDML file;</li>
<li>
<strong>AsDendrogram()</strong> – plots the structure of an RDML object as a dendrogram;</li>
<li>
<strong>AsTable()</strong> – represents the data contained in an RDML object (except fluorescence data) as a <em>data.frame</em>;</li>
<li>
<strong>GetFData()</strong> – gets fluorescence data;</li>
<li>
<strong>SetFData()</strong> – sets fluorescence data to an RDML object;</li>
<li>
<strong>AsXML()</strong> – saves an RDML object as an RDML<sub>v.</sub>1.2 file.</li>
</ul>
</div>
</div>
<div id="vendors-and-software-packages-supporting-rdml" class="section level2">
<h2 class="hasAnchor">
<a href="#vendors-and-software-packages-supporting-rdml" class="anchor"></a>Vendors and software packages supporting RDML</h2>
<p>The RDML format is supported by Bio-Rad (CFX 96 and CFX 384), Life Technologies (StepOne, ViiA7, QuantStudio) and Roche (LightCycler~96) thermocycler systems. In addition, several software packages (e.g., <strong>Primer3Plus</strong>, <strong>RDML-Ninja</strong>, <strong>qBase+</strong>) exist, which support RDML <span class="citation">(Untergasser et al. 2007; Hellemans et al. 2007; J M Ruijter et al. 2009; Pabinger et al. 2014; Jan M Ruijter et al. 2015)</span>.</p>
</div>
<div id="structure-of-the-rdml-package" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-the-rdml-package" class="anchor"></a>Structure of the <strong>RDML</strong> package</h2>
<p>The structure of the <strong>RDML</strong> package mimics the RDML format and provides several <a href="https://CRAN.R-project.org/package=R6/vignettes/Introduction.html">R6</a> classes that correspond to <a href="http://rdml.org/files.php?v=1.2">RDML v1.2</a> format types. All major manipulations with RDML data can be done by a class called <strong>RDML</strong> through its public methods:</p>
<ul>
<li>
<code>$new()</code> – creates new <strong>RDML</strong> object (empty or from specified RDML file)</li>
<li>
<code>$AsTable()</code> – represents data contained in <strong>RDML</strong> object (except fluorescent data) as data.frame.</li>
<li>
<code>$GetFData()</code> – gets fluorescent data.</li>
<li>
<code>$SetFData()</code> – sets fluorescent data to <strong>RDML</strong> object.</li>
<li>
<code>$AsDendrogram()</code> – represents structure of <strong>RDML</strong> object as dendrogram.</li>
</ul>
<div id="opening-and-observing-rdml-file" class="section level3">
<h3 class="hasAnchor">
<a href="#opening-and-observing-rdml-file" class="anchor"></a>Opening and observing RDML file</h3>
<p>In this section, we will use the built-in RDML example file <code>lc96_bACTXY.rdml</code>. This file was obtained during the measurement of human DNA concentration by a <em>LightCycler 96</em> (Roche) and the <em>XY-Detect</em> kit (Syntol, Russia).</p>
<p>To open the <code>lc96_bACTXY.rdml</code> file, we have to create a new <strong>RDML</strong> object with its class initializer – <code>$new()</code> and the file name as parameter <code>filename</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/lc96_bACTXY.rdml"</span>, <span class="dt">package =</span> <span class="st">"RDML"</span>)
lc96 &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dt">filename =</span> filename)</code></pre></div>
<p>Next we can check the structure of our new object – <code>lc96</code> by printing it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lc96
<span class="co">#&gt;  dateMade: 2014-08-19T11:25:48</span>
<span class="co">#&gt;  dateUpdated: 2014-08-27T12:06:21</span>
<span class="co">#&gt;  id: [Roche Diagnostics]</span>
<span class="co">#&gt;  experimenter: []</span>
<span class="co">#&gt;  documentation: []</span>
<span class="co">#&gt;  dye: [FAM, Hex, Texas Red, Cy5]</span>
<span class="co">#&gt;  sample: [Sample 39, Sample 41, Sample 43, Sample 45, Sample 51, Sample 53, Sample 54, Sample 55, Sample 56, Sample 57, Sample 58]</span>
<span class="co">#&gt;  target: [30116ec1-44f6-4c9c-9c69-5d6f00226d4e, 69b0b5cd-591c-4012-a995-7a8b53861548, 7797a698-1b2d-4819-bf7d-1188f2c8ca7f, c16f36ee-8636-40d2-ae72-b00d3b2eb89d, bACT, X, Y, IPC]</span>
<span class="co">#&gt;  thermalCyclingConditions: [2f78ed33-724e-4a29-97e9-92296eb868e1]</span>
<span class="co">#&gt;  experiment: [ca1eb225-ecea-4793-9804-87bfbb45f81d]</span></code></pre></div>
<p>As a result we can see field names, and after <code>:</code></p>
<ul>
<li>names of the <code>R6</code> objects contained at this field after <code>~</code>,</li>
<li>contained values after <code>:</code>,</li>
<li>names of list elements enclosed in <code>[]</code>.</li>
</ul>
<p>The field names for all <strong>RDML</strong> package classes correspond to field names of RDML types as described at <a href="http://rdml.org/files.php?v=1.2" class="uri">http://rdml.org/files.php?v=1.2</a>.</p>
<p><img src="rdml_overview_1_2.png" alt="HTML5 Icon" style="width:663px;height:688px"></p>
<p>Image source: <a href="http://rdml.org/" class="uri">http://rdml.org/</a>.</p>
<p>For the base class <strong>RDML</strong> they are:</p>
<ul>
<li><code>dateMade</code></li>
<li><code>dateUpdated</code></li>
<li>
<code>id</code> – publisher and id to the RDML file.</li>
<li>
<code>experimenter</code> – contact details of the experimenter.</li>
<li>
<code>documentation</code> – these elements should be used if the same description applies to many samples, targets, or experiments.</li>
<li>
<code>dye</code> – information about a dye.</li>
<li>
<code>sample</code> – defined template solutions.</li>
<li>
<code>target</code> – defined PCR reactions.</li>
<li>
<code>thermalCyclingConditions</code> – cycling programs for PCR or to amplify cDNA.</li>
<li><code>experiment</code></li>
</ul>
<p>These fields can be divided by two parts:</p>
<div id="experiment-field" class="section level4">
<h4 class="hasAnchor">
<a href="#experiment-field" class="anchor"></a>Experiment field</h4>
<p>Contains one or more experiments with fluorescence data. Fluorescence data are stored at the <em>data</em> level of an experiment. For example, fluorescence data for reaction tube <em>45</em> and target <em>bACT</em> can be accessed with the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fdata &lt;-<span class="st"> </span>
<span class="st">  </span>lc96$
<span class="st">    </span>experiment$<span class="st">`</span><span class="dt">ca1eb225-ecea-4793-9804-87bfbb45f81d</span><span class="st">`</span>$
<span class="st">    </span>run$<span class="st">`</span><span class="dt">65aeb1ec-b377-4ef6-b03f-92898d47488b</span><span class="st">`</span>$
<span class="st">    </span>react$<span class="st">`</span><span class="dt">45</span><span class="st">`</span>$
<span class="st">    </span>data$bACT$
<span class="st">    </span>adp$fpoints <span class="co">#'adp' means amplification data points (qPCR)</span>
<span class="kw">head</span>(fdata)
<span class="co">#&gt;    cyc     tmp     fluor</span>
<span class="co">#&gt; 1:   1 68.0054 0.0782385</span>
<span class="co">#&gt; 2:   2 68.0429 0.0753689</span>
<span class="co">#&gt; 3:   3 68.0451 0.0736838</span>
<span class="co">#&gt; 4:   4 68.0525 0.0723196</span>
<span class="co">#&gt; 5:   5 68.0537 0.0717019</span>
<span class="co">#&gt; 6:   6 68.0538 0.0714182</span></code></pre></div>
<p>The structure of experiments can be visualized by plotting dendrograms. <a name="dendrogram"></a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lc96$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/dendrogram-1.png" width="576"></p>
<p>In this dendrogram <a href="#dendrogram">1</a> we can see that our file consists of one experiment and one run. Four targets, each with two sample types (<em>std</em> – standard, <em>unkn</em> – unknown), are part of the experiment. There is only qPCR data – <em>adp</em> in this experiment. Ten reactions (tubes) for standard type (<em>std</em>) and six reaction for the unknown (<em>unkn</em>) type. The total number of reactions can be more than the number of reactions on the plate because one tube can contain more than one target (e.g., multiplexing).</p>
</div>
<div id="additional-information-fields" class="section level4">
<h4 class="hasAnchor">
<a href="#additional-information-fields" class="anchor"></a>Additional information fields</h4>
<p>All fields other than <strong>experiment</strong>. This additional information can be referenced in other parts of the RDML file. For example, to access sample added to react <em>39</em> and get its quantity, we can use code like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ref &lt;-<span class="st"> </span>lc96$
<span class="st">          </span>experiment$<span class="st">`</span><span class="dt">ca1eb225-ecea-4793-9804-87bfbb45f81d</span><span class="st">`</span>$
<span class="st">          </span>run$<span class="st">`</span><span class="dt">65aeb1ec-b377-4ef6-b03f-92898d47488b</span><span class="st">`</span>$
<span class="st">          </span>react$<span class="st">`</span><span class="dt">39</span><span class="st">`</span>$
<span class="st">          </span>sample$id
sample &lt;-<span class="st"> </span>lc96$sample[[ref]]
sample$quantity$value
<span class="co">#&gt; NULL</span></code></pre></div>
</div>
</div>
<div id="copying-rdml-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#copying-rdml-objects" class="anchor"></a>Copying <strong>RDML</strong> objects</h3>
<p><strong>R6</strong> objects are environments, that’s why simple copying results in creating references to existing objects. Then modifying the copy leads to modification of the original object. To create a <em>real</em> copy of the object we have to use method <code>$clone(deep = TRUE)</code> provided by the <strong>R6</strong> class.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">id1 &lt;-<span class="st"> </span>idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"id_1"</span>)
id2 &lt;-<span class="st"> </span>id1
id3 &lt;-<span class="st"> </span>id1$<span class="kw">clone</span>(<span class="dt">deep =</span> <span class="ot">TRUE</span>)
id2$id &lt;-<span class="st"> "id_2"</span>
id3$id &lt;-<span class="st"> "id_3"</span>
<span class="kw">cat</span>(<span class="kw">sprintf</span>(<span class="st">"Original object</span><span class="ch">\t</span><span class="st">: %s ('id_1' bacame 'id_2')</span><span class="ch">\n</span><span class="st">Simple copy</span><span class="ch">\t\t</span><span class="st">: %s</span><span class="ch">\n</span><span class="st">Clone</span><span class="ch">\t\t\t</span><span class="st">: %s</span><span class="ch">\n</span><span class="st">"</span>,
            id1$id, id2$id, id3$id))
<span class="co">#&gt; Original object  : id_2 ('id_1' bacame 'id_2')</span>
<span class="co">#&gt; Simple copy      : id_2</span>
<span class="co">#&gt; Clone            : id_3</span></code></pre></div>
<p>From the example above, we can see that the modification of <code>id2</code> led to a modification of the original object <code>id1</code>, however modification of cloned object <code>id3</code> did not.</p>
</div>
<div id="modifying-rdml-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#modifying-rdml-objects" class="anchor"></a>Modifying <strong>RDML</strong> objects</h3>
<p>To modify the content of <strong>RDML</strong> objects, we can use fields as setters. These setters provide type safe modifications by input validation. In addition, setting lists of objects generates names of list elements.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Create 'real' copy of object
experiment &lt;-<span class="st"> </span>lc96$experiment$<span class="st">`</span><span class="dt">ca1eb225-ecea-4793-9804-87bfbb45f81d</span><span class="st">`</span>$<span class="kw">clone</span>(<span class="dt">deep =</span> <span class="ot">TRUE</span>)
## Try to set 'id' with wrong input type.
## Correct type 'idType' can be seen at error message.
<span class="kw">tryCatch</span>(experiment$id &lt;-<span class="st"> "exp1"</span>,
         <span class="dt">error =</span> function(e) <span class="kw">print</span>(e))
<span class="co">#&gt; &lt;simpleError in withCallingHandlers({    saveRDS(do.call(do.call, c(readRDS("/tmp/Rtmp3L8dEU/file22553424b6b4"),         list(envir = .GlobalEnv)), envir = .GlobalEnv), file = "/tmp/Rtmp3L8dEU/file225546d4690")}, error = function(e) {    saveRDS(list("error", e), file = paste0("/tmp/Rtmp3L8dEU/file225546d4690",         ".error"))    stop(e)}): Assertion on 'id' failed: Must have class 'idType', but has class 'character'.&gt;</span>

## Set 'id' with correct input type - 'idType'
experiment$id &lt;-<span class="st"> </span>idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"exp1"</span>)

## Similar operations for 'run'
run &lt;-<span class="st"> </span>experiment$run$<span class="st">`</span><span class="dt">65aeb1ec-b377-4ef6-b03f-92898d47488b</span><span class="st">`</span>$<span class="kw">clone</span>(<span class="dt">deep =</span> <span class="ot">TRUE</span>)
run$id &lt;-<span class="st"> </span>idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"run1"</span>)

## Replace original elements with modified
experiment$run &lt;-<span class="st"> </span><span class="kw">list</span>(run)
lc96$experiment &lt;-<span class="st"> </span><span class="kw">list</span>(experiment)</code></pre></div>
<p><a name="dendrogram_modified"></a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lc96$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/dendrogram_modified-1.png" width="576"></p>
<p>Then, we can observe in <a href="#dendrogram_modified">2</a> our modification with the <code>$AsDendrogram()</code> method.</p>
</div>
<div id="astable-method" class="section level3">
<h3 class="hasAnchor">
<a href="#astable-method" class="anchor"></a><code><a href="../reference/astable-function.html">AsTable()</a></code> method</h3>
<p>To obtain information about all fluorescence data in an RDML file (type of added sample, used target, starting quantity etc.) as a data.frame, we can use the <code>$AsTable()</code> method. By default, it provides such information as:</p>
<ul>
<li>
<code>fdata.name</code> – aggregated name for current fluorescence data. Default pattern is <code>position_sample_sample.type_target</code> (e.g., <em>D03_Sample 39_std_bACT</em>). This pattern can be modified by <code>name.pattern</code> argument.</li>
<li>
<code>exp.id</code> – experiment id (e.g., <em>exp1</em>).</li>
<li>
<code>run.id</code> – run id (e.g., <em>run1</em>).</li>
<li>
<code>react.id</code> – react (tube) id (e.g., <em>39</em>).</li>
<li>
<code>position</code> – react (tube) position (e.g., <em>D03</em>).</li>
<li>
<code>sample</code> – name of the added sample (e.g., <em>Sample 39</em>).</li>
<li>
<code>target</code> – detection target (e.g., <em>bACT</em>).</li>
<li>
<code>target.dyeId</code> – detection dye (e.g., <em>FAM</em>).</li>
<li>
<code>sample,type</code> – type of the added sample (e.g., <em>std</em>).</li>
<li>
<code>adp</code> – <code>TRUE</code> if contains qPCR data.</li>
<li>
<code>mdp</code> – <code>TRUE</code> if contains melting data.</li>
</ul>
<p>To add custom columns to the output <em>data.frame</em>, we should pass it as a named method argument by generating expressions. Values for default columns can be used by custom name patterns and new columns referring to their names. Thr next example shows how to use the <code>$AsTable()</code> method with a custom name pattern and additional columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tab &lt;-<span class="st"> </span>lc96$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>(
  <span class="co"># Custom name pattern 'position~sample~sample.type~target~dye'</span>
  <span class="dt">name.pattern =</span> <span class="kw">paste</span>(
             react$position,
             react$sample$id,
             private$.sample[[react$sample$id]]$type$value,
             data$tar$id,
             target[[data$tar$id]]$dyeId$id,
             <span class="dt">sep =</span> <span class="st">"~"</span>),
  <span class="co"># Custom column 'quantity' - starting quantity of added sample </span>
  <span class="dt">quantity =</span> {
    value  &lt;-<span class="st"> </span>sample[[react$sample$id]]$quantity$value
    if (<span class="kw">is.null</span>(value) ||<span class="st"> </span><span class="kw">is.na</span>(value)) <span class="ot">NULL</span>
    else value
  }
)
## Remove row names for compact printing
<span class="kw">rownames</span>(tab) &lt;-<span class="st"> </span><span class="ot">NULL</span>
<span class="kw">head</span>(tab)
<span class="co">#&gt;                       fdata.name exp.id run.id react.id position    sample</span>
<span class="co">#&gt; 1:     D03~Sample 39~std~IPC~Cy5   exp1   run1       39      D03 Sample 39</span>
<span class="co">#&gt; 2:       D03~Sample 39~std~X~Hex   exp1   run1       39      D03 Sample 39</span>
<span class="co">#&gt; 3: D03~Sample 39~std~Y~Texas Red   exp1   run1       39      D03 Sample 39</span>
<span class="co">#&gt; 4:    D03~Sample 39~std~bACT~FAM   exp1   run1       39      D03 Sample 39</span>
<span class="co">#&gt; 5:     D04~Sample 39~std~IPC~Cy5   exp1   run1       40      D04 Sample 39</span>
<span class="co">#&gt; 6:       D04~Sample 39~std~X~Hex   exp1   run1       40      D04 Sample 39</span>
<span class="co">#&gt;    target target.dyeId sample.type  adp   mdp</span>
<span class="co">#&gt; 1:    IPC          Cy5         std TRUE FALSE</span>
<span class="co">#&gt; 2:      X          Hex         std TRUE FALSE</span>
<span class="co">#&gt; 3:      Y    Texas Red         std TRUE FALSE</span>
<span class="co">#&gt; 4:   bACT          FAM         std TRUE FALSE</span>
<span class="co">#&gt; 5:    IPC          Cy5         std TRUE FALSE</span>
<span class="co">#&gt; 6:      X          Hex         std TRUE FALSE</span></code></pre></div>
<p>Here, the generated data.frame is also used as a query in <code>$GetFData()</code> and <code>$SetFData()</code> methods (see further sections).</p>
</div>
<div id="getting-fluorescence-data" class="section level3">
<h3 class="hasAnchor">
<a href="#getting-fluorescence-data" class="anchor"></a>Getting fluorescence data</h3>
<p>We can extract the fluorescence data in two ways:</p>
<ul>
<li>direct access to the data as described in the <em>Experiment field</em> subsection</li>
<li>using the special method <code>$GetFData()</code>
</li>
</ul>
<p>The advantage of <code>$GetFData()</code> is that it can combine fluorescence data from many plates into one data.frame. The major argument of this function is <code>request</code>, which defines fluorescence data to be extracted. This request is the output from the <code>$AsTable()</code> method and can be easily filtered by the <strong>dplyr</strong> <code>filter()</code> function. Furthermore cycle limits, the output <code>data.frame</code> format as well as data types (<code>fdata.type = 'adp'</code> for qPCR, <code>fdata.type = 'mdp'</code> for melting data) can by specified (see examples below).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw">library</span>(ggplot2)

## Prepare request to get only 'std' type samples
filtered.tab &lt;-<span class="st"> </span><span class="kw">filter</span>(tab,
                       sample.type ==<span class="st"> "std"</span>)

fdata &lt;-<span class="st"> </span>lc96$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(filtered.tab,
                       <span class="co"># long table format for usage with ggplot2</span>
                       <span class="dt">long.table =</span> <span class="ot">TRUE</span>)
<span class="kw">ggplot</span>(fdata, <span class="kw">aes</span>(cyc, fluor)) +
<span class="st">    </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">group =</span> fdata.name,
                  <span class="dt">color =</span> target))</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-9-1.png" width="576"></p>
<p>Our example curves are not background subtracted, as visible in the plot. To accomplish this, we use the <code>CPP()</code> function from the <a href="https://CRAN.R-project.org/package=chipPCR">chipPCR</a> package <span class="citation">(Rödiger, Burdukiewicz, and Schierack 2015)</span> as described in <span class="citation">Rödiger et al. (2015)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chipPCR)
tab &lt;-<span class="st"> </span>lc96$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>(
  <span class="co"># Custom name pattern 'position~sample~sample.type~target~run.id'</span>
  <span class="dt">name.pattern =</span> <span class="kw">paste</span>(
             react$position,
             react$sample$id,
             private$.sample[[react$sample$id]]$type$value,
             data$tar$id,
             run$id$id, <span class="co"># run id added to names</span>
             <span class="dt">sep =</span> <span class="st">"~"</span>))
## Get all fluorescence data
fdata &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(lc96$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(tab,
                                     <span class="co"># We don't need long table format for CPP()</span>
                                     <span class="dt">long.table =</span> <span class="ot">FALSE</span>))

fdata.cpp &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">cyc =</span> fdata[, <span class="dv">1</span>],
                   <span class="kw">apply</span>(fdata[, -<span class="dv">1</span>], <span class="dv">2</span>,
                         function(x) <span class="kw">CPP</span>(fdata[, <span class="dv">1</span>],
                                         x)$y))</code></pre></div>
<p>Now we have properly preprocessed data, which we will add to our object and use in the next section.</p>
</div>
<div id="setting-fluorescence-data" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-fluorescence-data" class="anchor"></a>Setting fluorescence data</h3>
<p>To define fluorescence data as an <strong>RDML</strong> object, we can use the <code>$SetFData()</code> method. It takes three arguments:</p>
<ul>
<li>
<code>fdata</code> – fluorescence data in <code>long.table = FALSE</code> format</li>
<li>
<code>request</code> – output from the <code><a href="../reference/astable-function.html">AsTable()</a></code> function, which is used to set data;</li>
<li>
<code>fdata.type</code> – <code>fdata.type = 'adp'</code> for qPCR, <code>fdata.type = 'mdp'</code> for melting data.</li>
</ul>
<p>In the following we will define preprocessed fluorescence data as the new run – <em>run1_cpp</em>. The subelements of RDML such as <em>experiment</em>, <em>run</em>, <em>react</em> and <em>data</em> that do not initially exist in an <strong>RDML</strong> object are created by SetFData automaticaly (read more at <em>Creating <strong>RDML</strong> from table</em> section).</p>
<blockquote>
<p>Note that the colnames in <code>fdata</code> and <code>fdata.name</code> in <code>request</code> have to be the same!</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tab$run.id &lt;-<span class="st"> "run.cpp"</span>
## Set fluorescence data from previous section
lc96$<span class="kw"><a href="../reference/setfdata-function.html">SetFData</a></span>(fdata.cpp,
              tab)

## View setted data
fdata &lt;-<span class="st"> </span>lc96$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(tab,
                       <span class="dt">long.table =</span> <span class="ot">TRUE</span>)
<span class="kw">ggplot</span>(fdata, <span class="kw">aes</span>(cyc, fluor)) +
<span class="st">    </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">group =</span> fdata.name,
                  <span class="dt">color =</span> target))</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
</div>
<div id="merging-rdml-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#merging-rdml-objects" class="anchor"></a>Merging <strong>RDML</strong> objects</h3>
<p>Merging <strong>RDML</strong> objects can be conducted by the <code><a href="../reference/MergeRDMLs.html">MergeRDMLs()</a></code> function. It takes a list of <strong>RDML</strong> objects and returns one single <strong>RDML</strong> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Load another built in RDML file
stepone &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="kw">paste0</span>(<span class="kw">path.package</span>(<span class="st">"RDML"</span>),
                           <span class="st">"/extdata/"</span>, <span class="st">"stepone_std.rdml"</span>))
## Merge it with our 'lc96' object
merged &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MergeRDMLs.html">MergeRDMLs</a></span>(<span class="kw">list</span>(lc96, stepone))
## View structure of new object
merged$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()
<span class="co">#&gt; Warning in self$AsTable(): fdata.name column has duplicates! Try another</span>
<span class="co">#&gt; name.pattern.</span></code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
</div>
<div id="saving-rdml-object-as-rdml-file" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-rdml-object-as-rdml-file" class="anchor"></a>Saving <strong>RDML</strong> object as RDML file</h3>
<p>To save <strong>RDML</strong> objects in RDML file v1.2 format, we can use the <code>$AsXML()</code> method where the <code>file.name</code> argument is the name of new RDML file. Without <code>file.name</code> given, the function returns an XML tree.</p>
<p><code>lc96$AsXML("lc96.rdml")</code></p>
<p>One can use <a href="https://sourceforge.net/projects/qpcr-ninja/">RDML-ninja</a> or the <a href="http://rdml.org/tools.php?validator">RDML validator</a> from the RMDL consortiun to validate RDML files created by the <strong>RDML</strong> package.</p>
</div>
<div id="creating-custom-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-custom-functions" class="anchor"></a>Creating custom functions</h3>
<p><strong>R6</strong> classes allow to add methods to existing classes by using the <code>$set()</code> method. Suppose that we want to add a method to preprocess all fluorescence data and calculate Cq values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">RDML$<span class="kw">set</span>(<span class="st">"public"</span>, <span class="st">"CalcCq"</span>,
         function() {
           <span class="kw">library</span>(chipPCR)
           fdata &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(self$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(
             self$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>()))
           fdata &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">cyc =</span> fdata[, <span class="dv">1</span>],
                          <span class="kw">apply</span>(fdata[, -<span class="dv">1</span>],
                                <span class="dv">2</span>,
                                function(x)
                                  <span class="co"># Data preprocessing</span>
                                  <span class="kw">CPP</span>(fdata[, <span class="dv">1</span>],
                                      x)$y)
                          )
           
           <span class="kw">apply</span>(fdata[, -<span class="dv">1</span>], <span class="dv">2</span>,
                 function(x) {
                   <span class="kw">tryCatch</span>(
                     <span class="co"># Calculate Cq</span>
                     <span class="kw">th.cyc</span>(fdata[, <span class="dv">1</span>], x,
                            <span class="dt">auto =</span> <span class="ot">TRUE</span>)@.Data[<span class="dv">1</span>],
                     <span class="dt">error =</span> function(e) <span class="ot">NA</span>)
                 })
         }
)

## Create new object with our advanced class
stepone &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="kw">paste0</span>(<span class="kw">path.package</span>(<span class="st">"RDML"</span>),
                           <span class="st">"/extdata/"</span>, <span class="st">"stepone_std.rdml"</span>))</code></pre></div>
<p>We can then apply our new method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">stepone$<span class="kw">CalcCq</span>()
<span class="co">#&gt;         A01_NTC_RNase P_ntc_RNase P         A02_NTC_RNase P_ntc_RNase P </span>
<span class="co">#&gt;                           13.801661                           12.964268 </span>
<span class="co">#&gt;         A03_NTC_RNase P_ntc_RNase P       A04_pop1_RNase P_unkn_RNase P </span>
<span class="co">#&gt;                           14.399910                           13.529380 </span>
<span class="co">#&gt;       A05_pop1_RNase P_unkn_RNase P       A06_pop1_RNase P_unkn_RNase P </span>
<span class="co">#&gt;                           14.609048                           14.160886 </span>
<span class="co">#&gt;       A07_pop2_RNase P_unkn_RNase P       A08_pop2_RNase P_unkn_RNase P </span>
<span class="co">#&gt;                           11.839277                           10.947218 </span>
<span class="co">#&gt;       B01_pop2_RNase P_unkn_RNase P B02_STD_RNase P_10000.0_std_RNase P </span>
<span class="co">#&gt;                           13.167509                           13.415532 </span>
<span class="co">#&gt; B03_STD_RNase P_10000.0_std_RNase P B04_STD_RNase P_10000.0_std_RNase P </span>
<span class="co">#&gt;                           13.363262                           14.082731 </span>
<span class="co">#&gt;  B05_STD_RNase P_5000.0_std_RNase P  B06_STD_RNase P_5000.0_std_RNase P </span>
<span class="co">#&gt;                           13.039183                           14.012804 </span>
<span class="co">#&gt;  B07_STD_RNase P_5000.0_std_RNase P  B08_STD_RNase P_2500.0_std_RNase P </span>
<span class="co">#&gt;                            9.790724                           12.981337 </span>
<span class="co">#&gt;  C01_STD_RNase P_2500.0_std_RNase P  C02_STD_RNase P_2500.0_std_RNase P </span>
<span class="co">#&gt;                           15.666928                           13.132126 </span>
<span class="co">#&gt;  C03_STD_RNase P_1250.0_std_RNase P  C04_STD_RNase P_1250.0_std_RNase P </span>
<span class="co">#&gt;                           13.905808                           14.974769 </span>
<span class="co">#&gt;  C05_STD_RNase P_1250.0_std_RNase P   C06_STD_RNase P_625.0_std_RNase P </span>
<span class="co">#&gt;                           13.880906                           12.850727 </span>
<span class="co">#&gt;   C07_STD_RNase P_625.0_std_RNase P   C08_STD_RNase P_625.0_std_RNase P </span>
<span class="co">#&gt;                           16.915642                           13.706484</span></code></pre></div>
</div>
<div id="creating-rdml-from-tables" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-rdml-from-tables" class="anchor"></a>Creating <strong>RDML</strong> from tables</h3>
<p><strong>RDML</strong> objects can not only be generated from files but also from user data contained in data.frames. For this, we have to create an empty <strong>RDML</strong> object, create a data.frame containing the data and set it by the <code>$SetFData()</code> method. Minimal required information (samples, targets, dyes) will be created from the data description.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">#### Create simulated data with AmpSim() from chipPCR package
## Cq for data to be generated
Cqs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">15</span>, <span class="dv">17</span>, <span class="dv">19</span>, <span class="dv">21</span>)
## PCR si,ulation will be 35 cycles
fdata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">cyc =</span> <span class="dv">1</span>:<span class="dv">35</span>)
for(Cq in Cqs) {
  fdata &lt;-<span class="st"> </span><span class="kw">cbind</span>(fdata,
                 <span class="kw">AmpSim</span>(<span class="dt">cyc =</span> <span class="dv">1</span>:<span class="dv">35</span>, <span class="dt">Cq =</span> Cq)[, <span class="dv">2</span>])
}
## Set names for fluorescence curves
<span class="kw">colnames</span>(fdata)[<span class="dv">2</span>:<span class="dv">5</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"c1"</span>, <span class="st">"c2"</span>, <span class="st">"c3"</span>, <span class="st">"c4"</span>)

## Create minimal description
descr &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">fdata.name =</span> <span class="kw">c</span>(<span class="st">"c1"</span>, <span class="st">"c2"</span>, <span class="st">"c3"</span>, <span class="st">"c4"</span>),
  <span class="dt">exp.id =</span> <span class="kw">c</span>(<span class="st">"exp1"</span>, <span class="st">"exp1"</span>, <span class="st">"exp1"</span>, <span class="st">"exp1"</span>),
  <span class="dt">run.id =</span> <span class="kw">c</span>(<span class="st">"run1"</span>, <span class="st">"run1"</span>, <span class="st">"run1"</span>, <span class="st">"run1"</span>),
  <span class="dt">react.id =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>),
  <span class="dt">sample =</span> <span class="kw">c</span>(<span class="st">"s1"</span>, <span class="st">"s1"</span>, <span class="st">"s2"</span>, <span class="st">"s2"</span>),
  <span class="dt">sample.type =</span> <span class="kw">c</span>(<span class="st">"unkn"</span>, <span class="st">"unkn"</span>, <span class="st">"unkn"</span>, <span class="st">"unkn"</span>),
  <span class="dt">target =</span> <span class="kw">c</span>(<span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene1"</span>, <span class="st">"gene2"</span>),
  <span class="dt">target.dyeId =</span> <span class="kw">c</span>(<span class="st">"FAM"</span>, <span class="st">"ROX"</span>, <span class="st">"FAM"</span>, <span class="st">"ROX"</span>),
  <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>
)

## Create empty RDML object
sim &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>()
## Add fluorescence data
sim$<span class="kw"><a href="../reference/setfdata-function.html">SetFData</a></span>(fdata, descr)

## Observe object
sim$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fdata &lt;-<span class="st"> </span>sim$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(sim$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>(),
                      <span class="dt">long.table =</span> <span class="ot">TRUE</span>)
<span class="kw">ggplot</span>(fdata, <span class="kw">aes</span>(cyc, fluor)) +
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">group =</span> fdata.name,
                <span class="dt">color =</span> target,
                <span class="dt">linetype =</span> sample))</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-15-2.png" width="576"></p>
</div>
<div id="creating-rdml-from-abi-7500-v-2-software" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-rdml-from-abi-7500-v-2-software" class="anchor"></a>Creating <strong>RDML</strong> from <em>ABI 7500 v.2</em> software</h3>
<p><strong>RDML</strong> objects can be created from <em>.eds</em> files generated by the <em>ABI 7500 v.2</em> or <em>StepOne</em> software (here, all analyses have to be conducted within the device software!).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/from_abi7500/sce.eds"</span>, <span class="dt">package =</span> <span class="st">"RDML"</span>)
abi &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dt">filename =</span> filename)
abi$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
</div>
<div id="creating-rdml-from--csv-files" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-rdml-from--csv-files" class="anchor"></a>Creating <strong>RDML</strong> from <em>.csv</em> files</h3>
<p><strong>RDML</strong> objects can be generated from <em>.csv</em> files. The files need to contain a first column named <strong>cyc</strong> for qPCR data or <strong>tmp</strong> for melting data. The remaining columns must contain the raw fluorescence signals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/from_tables/fdata.csv"</span>, <span class="dt">package =</span> <span class="st">"RDML"</span>)
csv &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dt">filename =</span> filename)
csv$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-17-1.png" width="576"></p>
</div>
<div id="creating-rdml-from--xls-or--xslx-files" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-rdml-from--xls-or--xslx-files" class="anchor"></a>Creating <strong>RDML</strong> from <em>.xls</em> or <em>.xslx</em> files</h3>
<p><strong>RDML</strong> objects can also be generated from <em>.xls</em> or <em>.xslx</em> file. The files need to contain a <em>description</em> data.frame at tab sheet <em>description</em> and qPCR data at tab sheet <em>adp</em>, and/or qPCR data at tab sheet <em>mdp</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/from_tables/table.xlsx"</span>, <span class="dt">package =</span> <span class="st">"RDML"</span>)
xslx &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dt">filename =</span> filename)
<span class="co">#&gt; Warning in value[[3L]](cond): NAs introduced by coercion</span>

<span class="co">#&gt; Warning in value[[3L]](cond): NAs introduced by coercion</span>
xslx$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-18-1.png" width="576"></p>
</div>
<div id="functional-style" class="section level3">
<h3 class="hasAnchor">
<a href="#functional-style" class="anchor"></a>Functional style</h3>
<p>To provide a functional programming style that is more convenient in <strong>R</strong>, the <strong>RDML</strong> class methods have function wrappers:</p>
<ul>
<li>
<code>obj$AsTable(...)</code> – <code><a href="../reference/astable-function.html">AsTable(obj, ...)</a></code>
</li>
<li>
<code>obj$SetFData(...)</code> – <code><a href="../reference/setfdata-function.html">SetFData(obj, ...)</a></code>
</li>
<li>
<code>obj$GetFData(...)</code> – <code><a href="../reference/GetFData-function.html">GetFData(obj, ...)</a></code>
</li>
<li>
<code>obj$AsDendrogram(...)</code> – <code><a href="../reference/AsDendrogram-function.html">AsDendrogram(obj, ...)</a></code>
</li>
</ul>
</div>
<div id="using" class="section level3">
<h3 class="hasAnchor">
<a href="#using" class="anchor"></a>Using</h3>
</div>
<div id="summary" class="section level3">
<h3 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h3>
<p>The <strong>RDML</strong> package provides classes and methods to work with RDML data generated by real-time quantitative PCR devices or enables creating RDML files from user generated data. Because classes of the <strong>RDML</strong> package are built with <strong>R6</strong>, they can be modified by adding custom methods and ensure type safe usage through input validation.</p>
</div>
</div>
<div id="creating-rdml-file-from-raw-data" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-rdml-file-from-raw-data" class="anchor"></a>Creating RDML file from raw data</h1>
<p>This vignette describes important step how to create <strong>RDML</strong> object from user provided data and calculate Cq values with custom <strong>RDML</strong> class method. We use a real-world example to demonstrate the use of the <strong>RDML</strong> package.</p>
<p>For this purpose we used data from the <em>VideoScan</em> platform. The <em>VideoScan</em> platform is a highly versatile fluorescence microscope based imaging platform for the real-time analysis of biomolecular interactions. <em>VideoScan</em> has been extended by a heating/cooling-unit, which can be used for DNA melting curve analysis <span class="citation">(Rödiger, Böhm, and Schimke 2013)</span>, real-time quantitative PCR <span class="citation">(Rödiger et al. 2013)</span> and quantitative isothermal amplification <span class="citation">(Spiess et al. 2015)</span>. In this example we focus on qPCR data obtained from the amplification of the human gene <em>mlc-2v</em> <span class="citation">(Sheikh, Lyon, and Chen 2015)</span>. The experimental details (e.g., primer sequences, amplification conditions) are described in <span class="citation">(Rödiger et al. 2013; George et al. 2016)</span>.</p>
<p>In detail, this dataset contains three amplification curves collected by three runs monitored with TaqMan qPCR chemistry. Similar to other experimental platforms has the <em>VideoScan</em> systems an output as <em>csv</em> file (comma separated value). In general, <em>csv</em> is a format that is sufficient for an in house usage. However, the exchange with other scientists is demanding, since experimental metadata is not consistently included in csv files. These are for example information about qPCR conditions, primers, experimentalist and so on as described in the MIQE and RDML standard documentations.</p>
<p>All <em>csv</em> raw data were transferred to the <strong>R</strong> environment and stored as <em>rda</em> files in the <strong>C54</strong> dataset after completion of the experiment as described elsewhere <span class="citation">(R Development Core Team 2012)</span>. The <strong>R</strong> environment provides numerous packages for the import of standard file formats, including formats from spread sheet applications (e.g., <em>XLS</em>, <em>XLSX</em>, <em>ODS</em>) <span class="citation">(Schutten 2014, <span class="citation">Warnes et al. (2015)</span>, <span class="citation">Dragulescu (2014)</span>)</span>, an ODBC connection <span class="citation">(Ripley and 2002) 2015)</span> or from the clipboard (<code>read.table("clipboard")</code>). There are several graphical user interfaces (GUIs) for <strong>R</strong>, which facilitate an easy import of raw data <span class="citation">(Valero-Mora and Ledesma 2012)</span>. For example <strong>RKWard</strong> <span class="citation">(Rödiger et al. 2012)</span> has import functionality for <em>csv</em>, <em>txt</em>, <em>spss</em> and <em>stata</em> files.</p>
<p>We will use the <strong>C54</strong> dataset directly from the <a href="https://CRAN.R-project.org/package=chipPCR">chipPCR</a> package <span class="citation">(Rödiger, Burdukiewicz, and Schierack 2015)</span>. The amplification curves have different number of cycles and one of them contains a missing value due to a sensor dropout. Thus, this data should be preprocessed before Cq calculation.</p>
<p><strong>R</strong> is a language with dynamic typing. This is helpful while scripting, but can lead to problematic debugging of more complicated workflows. <strong><em>R6</em></strong> classes provide type-safe interfaces to set data without access to the inner structure of objects so that all imported data can be validated. This option is highly useful when creating packages at an intermediate level for other packages (e.g., such an approach does not permit set type <strong><em>character</em></strong> in place of type <strong><em>integer</em></strong>). Furthermore, the inheritance of <strong><em>R6</em></strong> objects unifies the structure of the package and streamlines the extending of its capabilities (since the whole package is written around a single base class).</p>
<div id="initialization" class="section level2">
<h2 class="hasAnchor">
<a href="#initialization" class="anchor"></a>Initialization</h2>
<p>First of all we have to load all necessary packages and fluorescence data into our <strong>R</strong> session.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RDML)
<span class="kw">library</span>(chipPCR)
<span class="kw">library</span>(MBmca)
<span class="kw">library</span>(ggplot2)
<span class="kw">data</span>(C54)</code></pre></div>
<p>We can see that our data are arranged in data.frame with cycle numbers in the first column and fluorescence values in the other columns. Such data structure is ready to use with the <strong>RDML</strong> package and does not need any conversion.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(C54)
<span class="co">#&gt; 'data.frame':    56 obs. of  4 variables:</span>
<span class="co">#&gt;  $ Cycle: num  0 1 2 3 4 5 6 7 8 9 ...</span>
<span class="co">#&gt;  $ D1   : num  0.141 0.146 0.152 0.151 0.153 ...</span>
<span class="co">#&gt;  $ D2   : num  0.148 0.148 0.148 0.149 0.149 ...</span>
<span class="co">#&gt;  $ D3   : num  0.145 0.147 0.148 0.148 0.149 ...</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
dat &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Cycle =</span> <span class="kw">rep</span>(C54[, <span class="dv">1</span>], <span class="kw">ncol</span>(C54) -<span class="st"> </span><span class="dv">1</span>),
                  <span class="dt">Experiment =</span> <span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">colnames</span>(C54)[-<span class="dv">1</span>], function(i) <span class="kw">rep</span>(i, <span class="kw">nrow</span>(C54)))),
                  <span class="dt">refMFi =</span> <span class="kw">unlist</span>(C54[, -<span class="dv">1</span>]))
<span class="kw">levels</span>(dat[[<span class="st">"Experiment"</span>]]) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"D1 - Stock cDNA"</span>, <span class="st">"D2 - 1/10 cDNA"</span>, <span class="st">"D3 - 1/100 cDNA"</span>)

<span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="dt">x =</span> Cycle, <span class="dt">y =</span> refMFi, <span class="dt">colour =</span> Experiment)) +<span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
</div>
<div id="creating-custom-method-for-rdml-class" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-custom-method-for-rdml-class" class="anchor"></a>Creating custom method for <strong>RDML</strong> class</h2>
<p>The preprocessing and Cq calculation steps are held within a <strong>RDML</strong> object. To do this we shall add custom method to the <strong>RDML</strong> class and use this advanced class as <strong>RDML</strong> object initializator (data import).</p>
<p>As described previously <span class="citation">(Rödiger et al. 2015)</span>, we first preprocess the raw data with the <strong>CPP</strong> function from the <a href="https://CRAN.R-project.org/package=chipPCR">chipPCR</a> package. Next, these preprocessed data are stored in new runs within the same experiment. For Cq calculation we used the <strong>diffQ2</strong> function from the <a href="https://CRAN.R-project.org/package=MBmca">MBmca</a> <span class="citation">(Rödiger, Böhm, and Schimke 2013)</span> package. Note, there are other functions available within <strong>R</strong> to determine the Cq values. The <a href="https://CRAN.R-project.org/package=qpcR">qpcR</a> package is one of the most popular packages for such tasks.</p>
<p>Additional our method will have two arguments:</p>
<ul>
<li>vector <code>last.cycle</code> – cycle limits for every amplification curve;</li>
<li>list <code>bg.range</code> – background range for every amplification curve.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">RDML$<span class="kw">set</span>(<span class="st">"public"</span>, <span class="st">"ProcessVideoScan"</span>,
         function(last.cycle,
                  bg.range) {
           <span class="co"># Get curves description</span>
           tab &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(self$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>())
           <span class="co"># Get fluorescent point</span>
           dat &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(self$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(tab))
           <span class="co"># Give new names to runs</span>
           <span class="co"># Preprocess amplification curve raw data as described in </span>
           <span class="co"># R&amp;ouml;diger et al. (2015) Bioinformatics. Note that the dataset</span>
           <span class="co"># has different length of cycle numbers and missing values (NAs).</span>
           <span class="co"># The CPP function from the chipPCR package prepares the data for </span>
           <span class="co"># further analysis.</span>
           tab[, <span class="st">"run.id"</span>] &lt;-<span class="st"> </span><span class="kw">paste0</span>(tab[, <span class="st">"run.id"</span>], <span class="st">"_CPP"</span>)
           
           <span class="co"># Curves will be used one by one</span>
           for(i in <span class="dv">2</span>:<span class="kw">length</span>(dat)) {
             <span class="co"># Preprocess data</span>
             preprocessed &lt;-<span class="st"> </span><span class="kw">CPP</span>(dat[<span class="dv">1</span>:last.cycle[i -<span class="st"> </span><span class="dv">1</span>], <span class="dv">1</span>], 
                                 dat[<span class="dv">1</span>:last.cycle[i -<span class="st"> </span><span class="dv">1</span>], i],
                                 <span class="dt">trans =</span> <span class="ot">TRUE</span>,
                                 <span class="dt">bg.range =</span> bg.range[[i -<span class="st"> </span><span class="dv">1</span>]])[[<span class="st">"y.norm"</span>]]
             <span class="co"># Create data.frame with cycle number and preprocessed curve Then</span>
             <span class="co"># give name to fluorescence points columns as before preprocessing</span>
             dat_CPP &lt;-<span class="st"> </span><span class="kw">cbind</span>(dat[<span class="dv">1</span>:last.cycle[i -<span class="st"> </span><span class="dv">1</span>], <span class="dv">1</span>],
                              preprocessed)
             <span class="kw">colnames</span>(dat_CPP)[<span class="dv">2</span>] &lt;-<span class="st"> </span>tab$fdata.name[i -<span class="st"> </span><span class="dv">1</span>]
             <span class="co"># Set preprocessed data with new description (new run names)</span>
             self$<span class="kw"><a href="../reference/setfdata-function.html">SetFData</a></span>(dat_CPP, tab)
             <span class="co"># Calculate and set Cq</span>
             <span class="co"># Set Cq from second derivative maximum method as described in </span>
             <span class="co"># R&amp;ouml;diger et al. (2015) Bioinformatics for preprocessed data.</span>
             <span class="co"># The diffQ2 function from the MBmca package </span>
             <span class="co"># (R&amp;ouml;diger et al. (2013), The R Journal) was used to calculate the</span>
             <span class="co"># Cq values of each amplification curve.</span>
             cq &lt;-<span class="st"> </span><span class="kw">diffQ2</span>(dat_CPP, <span class="dt">inder =</span> <span class="ot">TRUE</span>)[[<span class="st">"xTm1.2.D2"</span>]][<span class="dv">1</span>]
             self$experiment[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"exp.id"</span>]]]$
<span class="st">               </span>run[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"run.id"</span>]]]$
<span class="st">               </span>react[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"react.id"</span>]]]$
<span class="st">               </span>data[[tab[i -<span class="st"> </span><span class="dv">1</span>, <span class="st">"target"</span>]]]$cq &lt;-<span class="st"> </span>cq
           }
         }, <span class="dt">overwrite =</span> <span class="ot">TRUE</span>
)</code></pre></div>
</div>
<div id="rdml-object-creation" class="section level2">
<h2 class="hasAnchor">
<a href="#rdml-object-creation" class="anchor"></a><strong>RDML</strong> object creation</h2>
<p>Now we can create an empty <strong>RDML</strong> object with our advanced class and set the fluorescence data into it. For setting the data we have to create a description (metadata) of the amplification curves, which contain:</p>
<ul>
<li>
<code>fdata.name</code> – curve name. Have to be as columns names in data.frame with fluorescence data.</li>
<li>
<code>exp.id</code> – experiment name.</li>
<li>
<code>run.id</code> – run name. In our case we have three runs.</li>
<li>
<code>react.id</code> – react ID. E.g., number of tube within a reaction plate.</li>
<li>
<code>sample</code> – sample name.</li>
<li>
<code>type</code> – sample type.</li>
<li>
<code>target</code> – target name. E.g., gene of interest name.</li>
<li>
<code>target.dyeID</code> – fluorescent dye used for detection.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create a data frame of metadata </span>
description &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">fdata.name =</span> <span class="kw">c</span>(<span class="st">"D1"</span>, <span class="st">"D2"</span>, <span class="st">"D3"</span>),
  <span class="dt">exp.id =</span> <span class="kw">c</span>(<span class="st">"exp1"</span>, <span class="st">"exp1"</span>, <span class="st">"exp1"</span>),
  <span class="dt">run.id =</span> <span class="kw">c</span>(<span class="st">"run1"</span>, <span class="st">"run2"</span>, <span class="st">"run3"</span>),
  <span class="dt">react.id =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>),
  <span class="dt">sample =</span> <span class="kw">c</span>(<span class="st">"D1 - Stock cDNA"</span>, <span class="st">"D2 - 1/10 cDNA"</span>, <span class="st">"D3 - 1/100 cDNA"</span>),
  <span class="dt">sample.type =</span> <span class="kw">c</span>(<span class="st">"unkn"</span>, <span class="st">"unkn"</span>, <span class="st">"unkn"</span>),
  <span class="dt">target =</span> <span class="kw">c</span>(<span class="st">"MLC-2v"</span>, <span class="st">"MLC-2v"</span>, <span class="st">"MLC-2v"</span>),
  <span class="dt">target.dyeId =</span> <span class="kw">c</span>(<span class="st">"Cy5"</span>, <span class="st">"Cy5"</span>, <span class="st">"Cy5"</span>),
  <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>
)

<span class="co"># Create an empty RDML object</span>
video.scan &lt;-<span class="st"> </span>RDML$<span class="kw"><a href="../reference/new-method.html">new</a></span>()

<span class="co"># Add fluorescence data and metadata to the RDML object from a given source.</span>
<span class="co"># For the sake of simplicity we use the C54 dataset from the chipPCR package.</span>
video.scan$<span class="kw"><a href="../reference/setfdata-function.html">SetFData</a></span>(C54, description)</code></pre></div>
</div>
<div id="setting-additional-information" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-additional-information" class="anchor"></a>Setting additional information</h2>
<p>As optional step, we can add some information about our experiment. Such as experimenter contacts, references to documentation, description of samples and targets, thermal cycling conditions, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Add experimentalist information</span>
video.scan$experimenter &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">list</span>(
    experimenterType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"SR"</span>),
      <span class="st">"Stefan"</span>,
      <span class="st">"Roediger"</span>,
      <span class="st">"stefan.roediger@b-tu.de"</span>
    ),
    experimenterType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"CD"</span>),
      <span class="st">"Claudia"</span>,
      <span class="st">"Deutschmann"</span>
    )
  )

<span class="co"># Add a reference to documentation</span>
video.scan$documentation &lt;-<span class="st"> </span><span class="kw">list</span>(
  documentationType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Roediger et al. 2013"</span>),
    <span class="kw">paste</span>(<span class="st">"A Highly Versatile Microscope Imaging Technology Platform for the Multiplex"</span>,
          <span class="st">"Real-Time Detection of Biomolecules and Autoimmune Antibodies. S. Roediger,"</span>,
          <span class="st">"P. Schierack, A. Boehm, J. Nitschke, I. Berger, U. Froemmel, C. Schmidt, M."</span>,
          <span class="st">"Ruhland, I. Schimke, D. Roggenbuck, W. Lehmann and C. Schroeder. Advances in"</span>,
          <span class="st">"Biochemical Bioengineering/Biotechnology. 133:33-74, 2013."</span>,
          <span class="st">"https://www.ncbi.nlm.nih.gov/pubmed/22437246"</span>)
  )
)

cdna &lt;-<span class="st"> </span>
<span class="st">  </span>cdnaSynthesisMethodType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    <span class="dt">enzyme =</span> <span class="st">"SuperScript II"</span>,
    <span class="dt">primingMethod =</span>
      primingMethodType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"oligo-dt"</span>),
    <span class="dt">dnaseTreatment =</span> <span class="ot">TRUE</span>
  )
video.scan$sample$<span class="st">`</span><span class="dt">D1 - Stock cDNA</span><span class="st">`</span>$description &lt;-<span class="st"> "Input stock cDNA was used undiluted (D1)"</span>
video.scan$sample$<span class="st">`</span><span class="dt">D1 - Stock cDNA</span><span class="st">`</span>$cdnaSynthesisMethod &lt;-<span class="st"> </span>cdna
video.scan$sample$<span class="st">`</span><span class="dt">D2 - 1/10 cDNA</span><span class="st">`</span>$description &lt;-<span class="st"> "1/1000 diluted in A. bidest"</span>
video.scan$sample$<span class="st">`</span><span class="dt">D2 - 1/10 cDNA</span><span class="st">`</span>$cdnaSynthesisMethod &lt;-<span class="st"> </span>cdna
video.scan$sample$<span class="st">`</span><span class="dt">D3 - 1/100 cDNA</span><span class="st">`</span>$description &lt;-<span class="st"> "1/1000000 diluted in A. bidest"</span>
video.scan$sample$<span class="st">`</span><span class="dt">D3 - 1/100 cDNA</span><span class="st">`</span>$cdnaSynthesisMethod &lt;-<span class="st"> </span>cdna

video.scan$target$<span class="st">`</span><span class="dt">MLC-2v</span><span class="st">`</span>$xRef &lt;-<span class="st"> </span><span class="kw">list</span>(
  xRefType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"uniprot"</span>,
               <span class="st">"P10916"</span>)
)
video.scan$target$<span class="st">`</span><span class="dt">MLC-2v</span><span class="st">`</span>$sequences &lt;-<span class="st"> </span>
<span class="st">  </span>sequencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    forwardPrimer &lt;-<span class="st"> </span>oligoType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      <span class="dt">sequence =</span> <span class="st">"ACAGGGATGGCTTCATTGAC"</span>),
    reversePrimer &lt;-<span class="st"> </span>oligoType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      <span class="dt">sequence =</span> <span class="st">"ATGCGTTGAGAATGGTTTCC"</span>),
    probe1 &lt;-<span class="st"> </span>oligoType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
      <span class="dt">threePrimeTag =</span> <span class="st">"Atto647N"</span>,
      <span class="dt">sequence =</span> <span class="st">"CAGGGTCCGCTCCCTTAAGTTTCTCC"</span>,
      <span class="dt">fivePrimeTag =</span> <span class="st">"BHQ2"</span>)
  )

tcc &lt;-<span class="st"> </span>
<span class="st">  </span>thermalCyclingConditionsType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
    idType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>),
    <span class="dt">experimenter =</span> <span class="kw">list</span>(
      idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"SR"</span>),
      idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"CD"</span>)
    ),
    <span class="dt">step =</span> 
      <span class="kw">list</span>(
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">1</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dv">95</span>,
                                            <span class="dv">600</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">2</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dv">95</span>,
                                            <span class="dv">40</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">3</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="fl">58.5</span>,
                                            <span class="dv">90</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">4</span>,
          <span class="dt">temperature =</span> temperatureType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="fl">68.5</span>,
                                            <span class="dv">90</span>)
        ),
        stepType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(
          <span class="dt">nr =</span> <span class="dv">5</span>,
          <span class="dt">loop =</span> loopType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="dt">goto =</span> <span class="dv">2</span>,
                              <span class="dt">repeat.n =</span> <span class="dv">49</span>)
        )
      )
  )
video.scan$thermalCyclingConditions &lt;-<span class="st"> </span><span class="kw">list</span>(
  tcc
)

<span class="co">#add description of the experiment</span>
video.scan$experiment$exp1$description &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">paste</span>(<span class="st">"The aim was to amplify MLC-2v in the VideoScan and to monitor with a"</span>,
        <span class="st">"hydrolysis probe for MLC-2v. The primer sequences for MLC-2v were taken"</span>,
        <span class="st">"from Roediger et al. (2013). The amplification was detected in solution of"</span>,
        <span class="st">"the 1 HCU (see Roediger et al. 2013 for details). A 20 micro L PCR reaction"</span>,
        <span class="st">"was composed of 250 nM primer (forward and reverse), 1x Maxima Probe qPCR"</span>,
        <span class="st">"Master Mix (Fermentas), 1 micro L template (MLC-2v amplification product in"</span>,
        <span class="st">"different dilutions), 50 nM hydrolysis probe for MLC-2v and A."</span>,
        <span class="st">"bidest. During the amplification, fluorescence was measured at 59.5 degree"</span>,
        <span class="st">"Celsius. The Cy5 channel was used to monitor the MLC-2v specific hydrolysis"</span>,
        <span class="st">"probe. Input stock cDNA was used undiluted (D1). D2 was 1/1000 and D3"</span>,
        <span class="st">"1/1000000 diluted in A. bidest. The D1, D2, and D3 have different numbers"</span>,
        <span class="st">"measure points and D2 contains a missing value at cycle 37."</span>)
video.scan$experiment$exp1$run$run1$thermalCyclingConditions &lt;-<span class="st"> </span>idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>)
video.scan$experiment$exp1$run$run2$thermalCyclingConditions &lt;-<span class="st"> </span>idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>)
video.scan$experiment$exp1$run$run3$thermalCyclingConditions &lt;-<span class="st"> </span>idReferencesType$<span class="kw"><a href="../reference/new-method.html">new</a></span>(<span class="st">"Amplification"</span>)</code></pre></div>
</div>
<div id="processing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#processing-data" class="anchor"></a>Processing data</h2>
<p>After creating <strong>RDML</strong> object and setting data we can use our method <code>$ProcessVideoScan</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Process VideoScan data</span>
video.scan$<span class="kw">ProcessVideoScan</span>(<span class="dt">last.cycle =</span> <span class="kw">c</span>(<span class="dv">35</span>, <span class="dv">45</span>, <span class="dv">55</span>),
                            <span class="dt">bg.range =</span> <span class="kw">list</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8</span>),
                                           <span class="ot">NULL</span>,
                                           <span class="ot">NULL</span>))
<span class="co">#&gt; Warning in lm.coefs(x[BG], y[BG], method.reg): Chosen method lmrob failed to converge. </span>
<span class="co">#&gt;                    Performed linear regression.</span></code></pre></div>
<p>Then visualize the object with the <code>$AsDendrogram</code> method (runs with IDs that include <code>_CPP</code> in names, contain the preprocessed curves):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualize RDML object</span>
video.scan$<span class="kw"><a href="../reference/AsDendrogram-function.html">AsDendrogram</a></span>()</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-29-1.png" width="576"></p>
<p>And plot preprocessed curves indicating the Cq values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Visualise preprocessed data with Cq values as vertical dashed lines
<span class="co"># Add custom column that contains the calculated Cq</span>
tab &lt;-<span class="st"> </span>video.scan$<span class="kw"><a href="../reference/astable-function.html">AsTable</a></span>(<span class="dt">cq =</span> {
                 cq &lt;-<span class="st"> </span>data$cq
                 if (<span class="kw">is.null</span>(cq) ||<span class="st"> </span><span class="kw">is.na</span>(cq))
                   <span class="ot">NULL</span>
                 else
                   cq
               })
<span class="co"># Get preprocessed data in 'long.table' format</span>
dat &lt;-<span class="st"> </span>video.scan$<span class="kw"><a href="../reference/GetFData-function.html">GetFData</a></span>(tab[<span class="kw">grepl</span>(<span class="st">"_CPP"</span>, tab[[<span class="st">"run.id"</span>]]), ],
                           <span class="dt">long.table =</span> <span class="ot">TRUE</span>)
<span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="dt">x =</span> cyc, <span class="dt">y =</span> fluor)) +
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">group =</span> fdata.name, <span class="dt">color =</span> fdata.name),
            <span class="dt">size =</span> <span class="fl">1.2</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="kw">aes</span>(<span class="dt">xintercept =</span> cq, <span class="dt">color =</span> fdata.name),
             <span class="dt">linetype =</span> <span class="st">"longdash"</span>) +
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="st">"Cycle"</span>) +
<span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="st">"Fluorescence"</span>) +
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Amplification of human MLC-2v</span><span class="ch">\n</span><span class="st">VideoScan HCU"</span>)</code></pre></div>
<p><img src="RDML_files/figure-html/unnamed-chunk-30-1.png" width="576"></p>
<p>However, there are further <strong>R</strong> packages on <a href="https://cran.r-project.org/">CRAN</a>, <a href="http://bioconductor.org/">Bioconductor</a> and <a href="https://github.com/">GitHub</a>, which can be used with the RDML package <span class="citation">(Rödiger et al. 2015)</span> to deal with non-detects <span class="citation">(McCall et al. 2014)</span>, normalization <span class="citation">(Perkins et al. 2012)</span>, expression analysis <span class="citation">(Dvinge and Bertone 2009, matz_no_2013)</span> and periodicity in qPCR systems <span class="citation">(Spiess et al. 2016)</span>.</p>
</div>
<div id="export-to-file" class="section level2">
<h2 class="hasAnchor">
<a href="#export-to-file" class="anchor"></a>Export to file</h2>
<p>To save our <strong>RDML</strong> object as file we can use <code>$AsXML()</code> method with file name as argument:</p>
<p><code>video.scan$AsXML("filename.RDML")</code></p>
<p><em>Note that currently due to limits of <code>XML</code> package it is rather slow operation!!!</em></p>
<p>A XML files can be compressed to save space. The <strong>RMDL</strong> package uses the <code>zip()</code> function from the <strong>utils</strong> package with default values to compress the RDML file. Therefore, the compression mode depends on the operating system. The <strong>RDML</strong> function <code>AsXML()</code> without an argument returns one string, which contains the XML data. This can be used to compress it with any other compression algorithm.</p>
</div>
<div id="benchmark-of-the-rdml-generation" class="section level2">
<h2 class="hasAnchor">
<a href="#benchmark-of-the-rdml-generation" class="anchor"></a>Benchmark of the RDML generation</h2>
<p>The XML format becomes less efficient to work with at larger file sizes and complex structured information (e.g., matrices). Therefore, a limitation of RMDL is the process of the file generation. This characteristic is inherited by the <strong>RDML</strong> package. Speed is an issue in high-throughput applications. To give the user a foundation for a decision support if the implementation of the RMDL format is advisable, we performed a benchmark. We used the <strong>microbenchmark</strong> package <span class="citation">(Olaf Mersmann and Friedman 2014, <span class="citation">Radford (2014)</span>)</span> to assess the time required for generation of RDML files. The simulation included datasets with ranging from 1 to 1000 entries per file.</p>
</div>
<div id="dealing-with-other-data-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#dealing-with-other-data-formats" class="anchor"></a>Dealing with other data formats</h2>
<p>Researchers often collect gene expression data from more than one laboratory and and are thus required to analyze and aggregate many data sets. Adaptable data management - also known as “adaptive informatics” - is relevant in cases where data from different omics approaches and assays (e.g., flow cytometry, digital PCR, NGS) need to be merged <span class="citation">(Baker 2012)</span>. As RDML is based on XML, it is possible to converge the files with other formats such as HDF, <span class="citation">(Millard et al. 2011)</span> which is supported by <strong>R</strong> <span class="citation">(Fischer and Pau 2015)</span>. This enables extended data storage and analysis, but also a higher level of experimental data management.</p>
</div>
<div id="conclusion" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h2>
<p>In this example we used raw data from the <em>VideoScan</em> platform to demonstrate the creation of an <em>RDML</em> file via the <strong>RDML</strong> package. This procedure can be easily adapted to other experiments and devices and gives the researcher a tool to create a file in a standardized format for convenient data exchange. In contrast to simple <em>csv</em> files contains this file additional metadata in a defined structure. This in turn makes the data available to other scientists for reproducible research.</p>
</div>
<div id="benchmark" class="section level2">
<h2 class="hasAnchor">
<a href="#benchmark" class="anchor"></a>Benchmark</h2>
<p>The RDML format may also store data from high-throughput technologies such as digital PCR or 384-well qPCR. To make the RDML package the most appropriate tool for these applications, we optimized the saving functionality of the package. Hence, the RDML package is able to save even large numbers of qPCR experiments in reasonable time.</p>
<p>We benchmarked the time needed for saving RDML object with different numbers of experiments on two desktop machines:</p>
<ol style="list-style-type: decimal">
<li><p>Windows 7 x64 (build 7601) Service Pack 1 on Intel Core i5 (2,40 GHz) with 4,00 GB RAM. R version 3.2.1 (2015-06-18).</p></li>
<li><p>Kubuntu 15.04 (Kernel 3.19.0-27-generic) Intel(R) Pentium(R) CPU 997 @ 1.60GHz with 4,00 GB RAM. R version 3.1.2 (2014-10-31).</p></li>
</ol>
<p>In the case “1.” we used R version 3.2.1 (2015-06-18) and in the case “2.” we used R version 3.1.2 (2014-10-31)</p>
<p>To warrant fairest comparison and taking into account factors like garbage collection, we repeated the procedure 100 times.</p>
<p><img src="RDML_files/figure-html/unnamed-chunk-20-1.png" width="576"></p>
<table class="table">
<caption>Results of benchmark.</caption>
<thead><tr class="header">
<th align="right">Number of experiments</th>
<th align="right">Median time [s]</th>
<th align="left">Operating System</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">0.0525600</td>
<td align="left">Windows</td>
</tr>
<tr class="even">
<td align="right">8</td>
<td align="right">0.1965201</td>
<td align="left">Windows</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="right">0.4882732</td>
<td align="left">Windows</td>
</tr>
<tr class="even">
<td align="right">30</td>
<td align="right">0.7770339</td>
<td align="left">Windows</td>
</tr>
<tr class="odd">
<td align="right">32</td>
<td align="right">0.8454795</td>
<td align="left">Windows</td>
</tr>
<tr class="even">
<td align="right">50</td>
<td align="right">1.4801968</td>
<td align="left">Windows</td>
</tr>
<tr class="odd">
<td align="right">96</td>
<td align="right">3.8208417</td>
<td align="left">Windows</td>
</tr>
<tr class="even">
<td align="right">100</td>
<td align="right">4.0498602</td>
<td align="left">Windows</td>
</tr>
<tr class="odd">
<td align="right">300</td>
<td align="right">25.6566066</td>
<td align="left">Windows</td>
</tr>
<tr class="even">
<td align="right">384</td>
<td align="right">39.9697519</td>
<td align="left">Windows</td>
</tr>
<tr class="odd">
<td align="right">765</td>
<td align="right">142.1440033</td>
<td align="left">Windows</td>
</tr>
<tr class="even">
<td align="right">1000</td>
<td align="right">239.7105810</td>
<td align="left">Windows</td>
</tr>
<tr class="odd">
<td align="right">1</td>
<td align="right">0.0665418</td>
<td align="left">Unix</td>
</tr>
<tr class="even">
<td align="right">8</td>
<td align="right">0.2468777</td>
<td align="left">Unix</td>
</tr>
<tr class="odd">
<td align="right">20</td>
<td align="right">0.6101219</td>
<td align="left">Unix</td>
</tr>
<tr class="even">
<td align="right">30</td>
<td align="right">0.9658560</td>
<td align="left">Unix</td>
</tr>
<tr class="odd">
<td align="right">32</td>
<td align="right">1.0468907</td>
<td align="left">Unix</td>
</tr>
<tr class="even">
<td align="right">50</td>
<td align="right">1.8240828</td>
<td align="left">Unix</td>
</tr>
<tr class="odd">
<td align="right">96</td>
<td align="right">4.5716005</td>
<td align="left">Unix</td>
</tr>
<tr class="even">
<td align="right">100</td>
<td align="right">4.8454858</td>
<td align="left">Unix</td>
</tr>
<tr class="odd">
<td align="right">300</td>
<td align="right">29.0860889</td>
<td align="left">Unix</td>
</tr>
<tr class="even">
<td align="right">384</td>
<td align="right">44.9543657</td>
<td align="left">Unix</td>
</tr>
<tr class="odd">
<td align="right">765</td>
<td align="right">159.6792129</td>
<td align="left">Unix</td>
</tr>
<tr class="even">
<td align="right">1000</td>
<td align="right">266.5159046</td>
<td align="left">Unix</td>
</tr>
</tbody>
</table>
<p>The results above indicate that, thanks to the RDML package, an average laboratory workstation is able to save large RDML files in reasonable time (around 4 minutes for 1000 experiments).</p>
</div>
<div id="validation-of-rdml-files-created-by-the-rdml-package" class="section level2">
<h2 class="hasAnchor">
<a href="#validation-of-rdml-files-created-by-the-rdml-package" class="anchor"></a>Validation of RDML files created by the RDML package</h2>
<p>The correct structure of an RDML file is an important prerequisite to perform reproducible analysis. Unfortunately, the created files can not be properly tested to date because the official RDML validator <a href="http://rdml.org/tools.php">rdml.org</a> is not online from time to time. However, there is a solution to validate the files by the alternate editor - <em>RDML-Ninja</em> <span class="citation">(Jan M Ruijter et al. 2015)</span>.</p>
</div>
<div id="rdmledit-web-server" class="section level2">
<h2 class="hasAnchor">
<a href="#rdmledit-web-server" class="anchor"></a>RDMLedit web-server</h2>
<p>The <a href="RDMLedit%20web-server">http://shtest.evrogen.net/rdmlEdit/</a> is a web tool based on the functionalities of the RDML package. Aside from basic functionalities like data import, modification and export, RDMLedit also produces visualisations of qPCR data. In addition, RDMLedit can also merge multiple RDML files.</p>
<div id="local-deployment-of-rdmledit" class="section level3">
<h3 class="hasAnchor">
<a href="#local-deployment-of-rdmledit" class="anchor"></a>Local deployment of RDMLedit</h3>
<p>As a <a href="shiny%20application">https://shiny.rstudio.com/</a>, RDMLedit may be also deployed locally. It requires an installed R and RDML package. To initialize, the user has to open R and run the following command: <strong>rdmlEdit()</strong>.</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-baker_quantitative_2012">
<p>Baker, Monya. 2012. “Quantitative Data: Learning to Share.” <em>Nature Methods</em> 9 (1): 39–41. doi:<a href="https://doi.org/10.1038/nmeth.1815">10.1038/nmeth.1815</a>.</p>
</div>
<div id="ref-Chang_R6">
<p>Chang, Winston. 2015. <em>R6: Classes with Reference Semantics</em>. <a href="https://CRAN.R-project.org/package=R6" class="uri">https://CRAN.R-project.org/package=R6</a>.</p>
</div>
<div id="ref-dragulescu_xlsx_2014">
<p>Dragulescu, Adrian A. 2014. <em>Xlsx: Read, Write, Format Excel 2007 and Excel 97/2000/XP/2003 Files</em>. <a href="https://CRAN.R-project.org/package=xlsx" class="uri">https://CRAN.R-project.org/package=xlsx</a>.</p>
</div>
<div id="ref-dvinge_htqpcr:_2009">
<p>Dvinge, Heidi, and Paul Bertone. 2009. “HTqPCR: High-Throughput Analysis and Visualization of Quantitative Real-Time PCR Data in R.” <em>Bioinformatics</em> 25 (24): 3325–6. doi:<a href="https://doi.org/10.1093/bioinformatics/btp578">10.1093/bioinformatics/btp578</a>.</p>
</div>
<div id="ref-Fischer_HDF5">
<p>Fischer, Bernd, and Gregoire Pau. 2015. “rhdf5: HDF5 Interface to R.” <em>R Package Version 2.10.0</em>. <a href="https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html" class="uri">https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html</a>.</p>
</div>
<div id="ref-george_development_2016">
<p>George, Sandra, Stefan Rödiger, Christian Schröder, Michael Knaut, and Jan-Heiner Küpper. 2016. “Development of Multiplex PCR Systems for Expression Profiling of Human Cardiomyocytes Induced to Proliferate by Lentivirus Transduction of Upcyte Genes.” <em>Journal of Cellular Biotechnology</em> 2 (1): 35–55. doi:<a href="https://doi.org/10.3233/JCB-15025">10.3233/JCB-15025</a>.</p>
</div>
<div id="ref-hellemans_2007">
<p>Hellemans, Jan, Geert Mortier, Anne De Paepe, Frank Speleman, and Jo Vandesompele. 2007. “qBase Relative Quantification Framework and Software for Management and Automated Analysis of Real-Time Quantitative PCR Data.” <em>Genome Biology</em> 8 (2): R19. doi:<a href="https://doi.org/10.1186/gb-2007-8-2-r19">10.1186/gb-2007-8-2-r19</a>.</p>
</div>
<div id="ref-Lang_XML">
<p>Lang, Duncan Temple, and the CRAN Team. 2015. <em>XML: Tools for Parsing and Generating XML Within R and S-Plus</em>. <a href="https://CRAN.R-project.org/package=XML" class="uri">https://CRAN.R-project.org/package=XML</a>.</p>
</div>
<div id="ref-lefever_rdml_2009">
<p>Lefever, Steve, Jan Hellemans, Filip Pattyn, Daniel R. Przybylski, Chris Taylor, René Geurts, Andreas Untergasser, and Jo Vandesompele. 2009. “RDML: Structured Language and Reporting Guidelines for Real-Time Quantitative PCR Data.” <em>Nucleic Acids Research</em> 37 (7): 2065–9. <a href="https://nar.oxfordjournals.org/content/37/7/2065" class="uri">https://nar.oxfordjournals.org/content/37/7/2065</a>.</p>
</div>
<div id="ref-mccall_non-detects_2014">
<p>McCall, Matthew N., Helene R. McMurray, Hartmut Land, and Anthony Almudevar. 2014. “On Non-Detects in qPCR Data.” <em>Bioinformatics</em> 30 (16): 2310–6. doi:<a href="https://doi.org/10.1093/bioinformatics/btu239">10.1093/bioinformatics/btu239</a>.</p>
</div>
<div id="ref-millard_adaptive_2011">
<p>Millard, Bjorn L., Mario Niepel, Michael P. Menden, Jeremy L. Muhlich, and Peter K. Sorger. 2011. “Adaptive Informatics for Multifactorial and High-Content Biological Data.” <em>Nature Methods</em> 8 (6): 487–92.</p>
</div>
<div id="ref-xml_book">
<p>Nolan, Deborah, and Duncan Temple Lang. 2014. <em>XML and Web Technologies for Data Sciences with R</em>. Springer. <a href="https://www.springer.com/statistics/computational+statistics/book/978-1-4614-7899-7" class="uri">https://www.springer.com/statistics/computational+statistics/book/978-1-4614-7899-7</a>.</p>
</div>
<div id="ref-Mersmann_2014_microbenchmark">
<p>Olaf Mersmann, Rainer Hurling, Claudia Beleites, and Ari Friedman. 2014. <em>microbenchmark: Accurate Timing Functions</em>. <a href="https://CRAN.R-project.org/package=microbenchmark" class="uri">https://CRAN.R-project.org/package=microbenchmark</a>.</p>
</div>
<div id="ref-pabinger_2014">
<p>Pabinger, Stephan, Stefan Rödiger, Albert Kriegner, Klemens Vierlinger, and Andreas Weinhäusel. 2014. “A Survey of Tools for the Analysis of Quantitative PCR (qPCR) Data.” <em>Biomolecular Detection and Quantification</em> 1 (1): 23–33.</p>
</div>
<div id="ref-perkins_readqpcr_2012">
<p>Perkins, James R., John M. Dawes, Steve B. McMahon, David LH Bennett, Christine Orengo, and Matthias Kohl. 2012. “ReadqPCR and NormqPCR: R Packages for the Reading, Quality Checking and Normalisation of RT-qPCR Quantification Cycle (Cq) Data.” <em>BMC Genomics</em> 13 (1): 296. doi:<a href="https://doi.org/10.1186/1471-2164-13-296">10.1186/1471-2164-13-296</a>.</p>
</div>
<div id="ref-RDCT2010c">
<p>R Development Core Team. 2012. <em>R Data Import/Export</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>.</p>
</div>
<div id="ref-Radford_2014_microbenchmark">
<p>Radford, Neal. 2014. <em>Inaccurate Results from Microbenchmark</em>. <a href="https://web.archive.org/web/20140823221902/https://www.r-bloggers.com/inaccurate-results-from-microbenchmark/" class="uri">https://web.archive.org/web/20140823221902/https://www.r-bloggers.com/inaccurate-results-from-microbenchmark/</a>.</p>
</div>
<div id="ref-Ren_rlist">
<p>Ren, Kun. 2015. <em>rlist: A Toolbox for Non-Tabular Data Manipulation</em>. <a href="https://CRAN.R-project.org/package=rlist" class="uri">https://CRAN.R-project.org/package=rlist</a>.</p>
</div>
<div id="ref-ripley_rodbc_2015">
<p>Ripley, Brian, and Michael Lapsley (1999 to Oct 2002). 2015. <em>RODBC: ODBC Database Access</em>. <a href="https://CRAN.R-project.org/package=RODBC" class="uri">https://CRAN.R-project.org/package=RODBC</a>.</p>
</div>
<div id="ref-roediger_RJ_2013">
<p>Rödiger, Stefan, Alexander Böhm, and Ingolf Schimke. 2013. “Surface Melting Curve Analysis with R.” <em>The R Journal</em> 5 (2): 37–53. <a href="https://journal.r-project.org/archive/2013-2/roediger-bohm-schimke.pdf%20">
https://journal.r-project.org/archive/2013-2/roediger-bohm-schimke.pdf</a>.</p>
</div>
<div id="ref-roediger2015chippcr">
<p>Rödiger, Stefan, Michał Burdukiewicz, and Peter Schierack. 2015. “chipPCR: an R package to pre-process raw data of amplification curves.” <em>Bioinformatics</em> 31 (17): 2900–2902. <a href="https://bioinformatics.oxfordjournals.org/content/31/17/2900.abstract" class="uri">https://bioinformatics.oxfordjournals.org/content/31/17/2900.abstract</a>.</p>
</div>
<div id="ref-roediger2015r">
<p>Rödiger, Stefan, Michał Burdukiewicz, Konstantin Blagodatskikh, and Peter Schierack. 2015. “R as an Environment for the Reproducible Analysis of Dna Amplification Experiments.” <em>The R Journal</em> 7 (2): 127–50.</p>
</div>
<div id="ref-roediger_rkward_2012">
<p>Rödiger, Stefan, Thomas Friedrichsmeier, Prasenjit Kapat, and Meik Michalke. 2012. “RKWard: A Comprehensive Graphical User Interface and Integrated Development Environment for Statistical Analysis with R.” <em>Journal of Statistical Software</em> 49 (9): 1–34. <a href="https://www.jstatsoft.org/article/view/v049i09" class="uri">https://www.jstatsoft.org/article/view/v049i09</a>.</p>
</div>
<div id="ref-roediger_highly_2013">
<p>Rödiger, Stefan, Peter Schierack, Alexander Böhm, Jörg Nitschke, Ingo Berger, Ulrike Frömmel, Carsten Schmidt, et al. 2013. “A highly versatile microscope imaging technology platform for the multiplex real-time detection of biomolecules and autoimmune antibodies.” <em>Advances in Biochemical Engineering/Biotechnology</em> 133: 35–74.</p>
</div>
<div id="ref-ruijter_amplification_2009">
<p>Ruijter, J M, C Ramakers, W M H Hoogaars, Y Karlen, O Bakker, M J B van den Hoff, and A F M Moorman. 2009. “Amplification Efficiency: Linking Baseline and Bias in the Analysis of Quantitative PCR Data.” <em>Nucleic Acids Research</em> 37 (6): e45.</p>
</div>
<div id="ref-rdml-ninja_2015">
<p>Ruijter, Jan M, Steve Lefever, Jasper Anckaert, Jan Hellemans, Michael W Pfaffl, Vladimir Benes, Stephen A Bustin, Jo Vandesompele, Andreas Untergasser, and others. 2015. “RDML-Ninja and RDMLdb for Standardized Exchange of qPCR Data.” <em>BMC Bioinformatics</em> 16 (1). BioMed Central Ltd: 197.</p>
</div>
<div id="ref-Schutten_2014_ods">
<p>Schutten, Gerrit-Jan. 2014. <em>readODS: Read ODS Files and Puts Them into Data Frames</em>. <a href="https://CRAN.R-project.org/package=readODS" class="uri">https://CRAN.R-project.org/package=readODS</a>.</p>
</div>
<div id="ref-sheikh_2015">
<p>Sheikh, Farah, Robert C. Lyon, and Ju Chen. 2015. “Functions of Myosin Light Chain-2 (MYL2) in Cardiac Muscle and Disease.” <em>Gene</em> 569 (1): 14–20.</p>
</div>
<div id="ref-spiess_impact_2014">
<p>Spiess, Andrej-Nikolai, Claudia Deutschmann, Michał Burdukiewicz, Ralf Himmelreich, Katharina Klat, Peter Schierack, and Stefan Rödiger. 2015. “Impact of Smoothing on Parameter Estimation in Quantitative DNA Amplification Experiments.” <em>Clinical Chemistry</em> 61 (2): 379–88. doi:<a href="https://doi.org/10.1373/clinchem.2014.230656">10.1373/clinchem.2014.230656</a>.</p>
</div>
<div id="ref-spiess_system-specific_2016">
<p>Spiess, Andrej-Nikolai, Stefan Rödiger, Michał Burdukiewicz, Thomas Volksdorf, and Joel Tellinghuisen. 2016. “System-Specific Periodicity in Quantitative Real-Time Polymerase Chain Reaction Data Questions Threshold-Based Quantitation.” <em>Scientific Reports</em> 6 (December): 38951. doi:<a href="https://doi.org/10.1038/srep38951">10.1038/srep38951</a>.</p>
</div>
<div id="ref-untergasser_2007">
<p>Untergasser, Andreas, Harm Nijveen, Xiangyu Rao, Ton Bisseling, René Geurts, and Jack A. M. Leunissen. 2007. “Primer3Plus, an Enhanced Web Interface to Primer3.” <em>Nucleic Acids Research</em> 35 (suppl 2): W71–W74. doi:<a href="https://doi.org/10.1093/nar/gkm306">10.1093/nar/gkm306</a>.</p>
</div>
<div id="ref-Valero_2012">
<p>Valero-Mora, Pedro M., and Ruben Ledesma. 2012. “Graphical User Interfaces for R.” <em>Journal of Statistical Software</em> 49 (1): 1–8. <a href="https://www.jstatsoft.org/article/view/v049i01" class="uri">https://www.jstatsoft.org/article/view/v049i01</a>.</p>
</div>
<div id="ref-Warnes_2015_gdata">
<p>Warnes, Gregory R., Ben Bolker, Gregor Gorjanc, Gabor Grothendieck, Ales Korosec, Thomas Lumley, Don MacQueen, Arni Magnusson, Jim Rogers, and others. 2015. <em>gdata: Various R Programming Tools for Data Manipulation</em>. <a href="https://CRAN.R-project.org/package=gdata" class="uri">https://CRAN.R-project.org/package=gdata</a>.</p>
</div>
<div id="ref-Wickham_plyr">
<p>Wickham, Hadley. 2011. “The Split-Apply-Combine Strategy for Data Analysis.” <em>Journal of Statistical Software</em> 40 (1): 1–29. <a href="https://www.jstatsoft.org/article/view/v040i01" class="uri">https://www.jstatsoft.org/article/view/v040i01</a>.</p>
</div>
<div id="ref-Wickham_assertthat">
<p>———. 2013. <em>assertthat: Easy Pre and Post Assertions.</em> <a href="https://CRAN.R-project.org/package=assertthat" class="uri">https://CRAN.R-project.org/package=assertthat</a>.</p>
</div>
<div id="ref-Wickham_tidyr">
<p>———. 2014. <em>tidyr: Easily Tidy Data with Spread() and Gather() Functions.</em> <a href="https://CRAN.R-project.org/package=tidyr" class="uri">https://CRAN.R-project.org/package=tidyr</a>.</p>
</div>
<div id="ref-Wickham_Francois_dplyr">
<p>Wickham, Hadley, and Romain Francois. 2015. <em>dplyr: A Grammar of Data Manipulation</em>. <a href="https://CRAN.R-project.org/package=dplyr" class="uri">https://CRAN.R-project.org/package=dplyr</a>.</p>
</div>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction-to-the-rdml-package">Introduction to the RDML package</a></li>
      <li><a href="#vendors-and-software-packages-supporting-rdml">Vendors and software packages supporting RDML</a></li>
      <li><a href="#structure-of-the-rdml-package">Structure of the <strong>RDML</strong> package</a></li>
      <li>
<a href="#creating-rdml-file-from-raw-data">Creating RDML file from raw data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#initialization">Initialization</a></li>
      <li><a href="#creating-custom-method-for-rdml-class">Creating custom method for <strong>RDML</strong> class</a></li>
      <li><a href="#rdml-object-creation"><strong>RDML</strong> object creation</a></li>
      <li><a href="#setting-additional-information">Setting additional information</a></li>
      <li><a href="#processing-data">Processing data</a></li>
      <li><a href="#export-to-file">Export to file</a></li>
      <li><a href="#benchmark-of-the-rdml-generation">Benchmark of the RDML generation</a></li>
      <li><a href="#dealing-with-other-data-formats">Dealing with other data formats</a></li>
      <li><a href="#conclusion">Conclusion</a></li>
      <li><a href="#benchmark">Benchmark</a></li>
      <li><a href="#validation-of-rdml-files-created-by-the-rdml-package">Validation of RDML files created by the RDML package</a></li>
      <li><a href="#rdmledit-web-server">RDMLedit web-server</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Konstantin A. Blagodatskikh, Stefan Roediger, Michal Burdukiewicz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
